---
title: "scRNAseq essential analysis with Seurat with sctransform"
author: "Zhuang Liu"
date: "10/10/2021 Updated: `r Sys.Date()`"
output: 
  html_document:
    toc: true #add a table of contents (TOC)
    toc_depth: 3 #headers' level in the toc
    toc_float: true #toc visible even when scrolled
    theme: lumen
    highlight: tango
    code_folding: show
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align='center', fig.width=12, fig.height = 8)
```

```{r warning=FALSE, message=FALSE}
library(Seurat)
library(tidyverse)
library(magrittr)
library(ggplot2)
library(patchwork)
library(sctransform)
library(cowplot)
library(harmony)
```

#####################################################
# Step 1: Load 10X filtered matrices into seurat
#####################################################
## 1.1 set up the Seurat objects
```{r eval=FALSE, warning=FALSE, message=FALSE}
#Prior to load the data, please make sure the gene feature names shoud use uppercases (especiall for mouse).
samples <- list.files("/Users/zhuliu/Desktop/scRNA_STseq/proj_10X_woundhealing/03_results/00-cellrange_counts/")
sm_folders <- paste0("/Users/zhuliu/Desktop/scRNA_STseq/proj_10X_woundhealing/03_results/00-cellrange_counts/", samples, "/outs/filtered_feature_bc_matrix")

sceList <- list()
for (i in seq_along(samples)) {
  tmp_sms <- samples[i]
  sceList[[tmp_sms]] <- CreateSeuratObject(
                                          counts = Read10X(sm_folders[1]),
                                          min.cells = 3, # filter this later
                                          min.features = 200, # filter this later
                                          project = tmp_sms)
  
}

# filtering by nCount and nFeatures per individual
filterCell <- function(combined){
  # calculate the quantile range
  count.feature.ls <- combined@meta.data[, c("nCount_RNA", "nFeature_RNA")]
  count.feature.ls %<>% map(log10) %>% map(~c(10^(mean(.x) + 3*sd(.x)), 10^(mean(.x) - 3*sd(.x))))

  # filter cells
  combined <- subset(combined, subset = nFeature_RNA > 200 & 
                     nFeature_RNA < count.feature.ls[[2]][1] & 
                     nCount_RNA < count.feature.ls[[1]][1]) 
  return(combined)
}
sceList %<>% map(filterCell)

hswound.com <- merge(sceList[[1]], 
                      y = c(sceList[[2]], sceList[[3]], sceList[[4]], sceList[[5]], sceList[[6]],
			                      sceList[[7]], sceList[[8]], sceList[[9]], sceList[[10]], sceList[[11]], sceList[[12]]), 
                      add.cell.ids = samples, 
                      project = "humanWounds")
hswound.com #
head(colnames(hswound.com)); tail(colnames(hswound.com))
#cell numbers of each sample
table(Idents(hswound.com))
rm(sceList) #remove individual seurat objects
```
## 1.2 add sample level metadata (Age, Sex, etc)
```{r eval=FALSE}
metadata <- readxl::read_xlsx("/Users/zhuliu/Desktop/scRNA_STseq/proj_10X_woundhealing/02_metadata/Metadata_STsc.xlsx") %>% slice(17:28)
metadata <- metadata %>% select(3:6,11) %>% mutate(User_ID = gsub("_","",User_ID)) %>% rename(ID=User_ID)
metadata <- hswound.com@meta.data %>% left_join(., metadata, by=c("orig.ident" = "ID"))
rownames(metadata) <- Cells(hswound.com)
hswound.com <- AddMetaData(hswound.com, metadata = metadata)
saveRDS(hswound.com, file = "../../00_Origin_objects/s0_original_seurat_object.rds")
```


Original cell numbers for each sample
```{r eval=TRUE, echo=FALSE, warning=FALSE, message=FALSE}
hswound.com <- readRDS("../../00_Origin_objects/s0_original_seurat_object.rds")
rownames(hswound.com)[grep("LOC105372576", rownames(hswound.com))]

hswound.com
table(hswound.com$orig.ident)
```

## 1.3 load file and filter out doublets for following quality control
```{r eval=FALSE}
#filter out the doublets
doublets <- list.files("../../01_Seurat_qc_Doublets", pattern = "^Shared_doublets_")
doublets.path <- paste0("../../01_Seurat_qc_Doublets/", doublets)
doublets_files <- lapply(doublets.path, FUN = function(x){
  data.table::fread(x)
})
names(doublets_files) <- gsub("(Shared_doublets_)|(\\.txt)", "", doublets) 
doublets_com <- do.call("rbind", doublets_files) %>% 
  mutate(Doublet = "Doublet")
rm(doublets, doublets.path, doublets_files)
table(doublets_com$sampleID)

metadata <- hswound.com@meta.data %>% rownames_to_column(var = "mapID") %>% 
  left_join(., doublets_com[,c(3:4)], by=c("mapID" = "mapID")) %>% 
  column_to_rownames(var = "mapID") %>% 
  mutate(Doublet = ifelse(is.na(Doublet), "Singlet", Doublet))

hswound.com <- AddMetaData(hswound.com, metadata = metadata)
#Extract the singlet cells
hswound.com <- subset(hswound.com, subset = Doublet == "Singlet")
table(hswound.com$Doublet)
rm(metadata, doublets_com)
saveRDS(hswound.com, file = "../00_Origin_objects/s0_original_seurat_object_noDoublets.rds") 
```
Cell numbers for each sample after removing doublets
```{r eval=TRUE, echo=FALSE, warning=FALSE, message=FALSE}
hswound.com <- readRDS("../../00_Origin_objects/s0_original_seurat_object_noDoublets.rds")
hswound.com
table(hswound.com$orig.ident)
```

#####################################################
# Step 2: Quality Control
#####################################################
## 2.1 calculate QC
```{r eval=TRUE, fig.height=4, fig.width=6}
#ratio of mitochondrial genes
hswound.com[["percent.mt"]] <- PercentageFeatureSet(object = hswound.com, pattern = "^MT-")
rownames(hswound.com)[grep("^MT-", rownames(hswound.com))]

#ratio of ribosomal genes
hswound.com[["percent.ribo"]] <- PercentageFeatureSet(object = hswound.com, pattern = "^RP[SL]")
rownames(hswound.com)[grep("^RP[SL]", rownames(hswound.com))]

#ratio of hemoglobin genes
hswound.com[["percent.hb"]] <- PercentageFeatureSet(object = hswound.com, pattern = "^HB[^(PES)]")
rownames(hswound.com)[grep("^HB[^(PES)]", rownames(hswound.com))]
rownames(hswound.com)[grep("^HB", rownames(hswound.com))]

#ratio of MALAT1 genes
hswound.com[["percent.malat1"]] <- PercentageFeatureSet(object = hswound.com, pattern = "MALAT1")

sd(hswound.com@meta.data$nFeature_RNA)
d <- density(hswound.com@meta.data$nFeature_RNA)
{plot(d)
polygon(d, col="red", border="blue")}
table(hswound.com@meta.data$nFeature_RNA > 500)

summary(hswound.com@meta.data$nCount_RNA)
d <- density(hswound.com@meta.data$nCount_RNA)
{plot(d)
polygon(d, col="red", border="blue")}
table(hswound.com@meta.data$nCount_RNA > 1000)
table(hswound.com@meta.data$nCount_RNA > 1000 & hswound.com@meta.data$nCount_RNA < 150000)

summary(hswound.com@meta.data$percent.mt)
d <- density(hswound.com@meta.data$percent.mt)
{plot(d)
polygon(d, col="red", border="blue")}
table(hswound.com@meta.data$percent.mt < 20)
table(hswound.com@meta.data$percent.mt < 15)
table(hswound.com@meta.data$percent.mt < 10)

summary(hswound.com@meta.data$percent.ribo)
summary(hswound.com@meta.data$percent.hb)
```
## 2.2 QC plot before filtering
```{r fig.align='center', fig.width=12, fig.height=12}
VlnPlot(
  hswound.com,
  features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.ribo", "percent.hb"),
  ncol = 2, pt.size=0.0001)
```

```{r fig.align='center', fig.height=8}
plot1 <- FeatureScatter(hswound.com, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "orig.ident") + RotatedAxis()
plot2 <- FeatureScatter(hswound.com, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by = "orig.ident") + RotatedAxis()
plot3 <- FeatureScatter(hswound.com, feature1 = "nCount_RNA", feature2 = "percent.ribo", group.by = "orig.ident") + RotatedAxis()
plot4 <- FeatureScatter(hswound.com, feature1 = "nCount_RNA", feature2 = "percent.hb", group.by = "orig.ident") + RotatedAxis()
plot1 + plot2 + plot3 + plot4 + plot_layout(ncol = 2, guides = 'collect')
```

```{r fig.align='center', fig.height=8}
plot5 <- FeatureScatter(hswound.com, feature1 = "percent.mt", feature2 = "percent.malat1", group.by = "orig.ident") + RotatedAxis()
plot6 <- FeatureScatter(hswound.com, feature1 = "nCount_RNA", feature2 = "percent.malat1", group.by = "orig.ident") + RotatedAxis()
plot7 <- FeatureScatter(hswound.com, feature1 = "nFeature_RNA", feature2 = "percent.malat1", group.by = "orig.ident") + RotatedAxis()
plot8 <- FeatureScatter(hswound.com, feature1 = "nFeature_RNA", feature2 = "percent.mt", group.by = "orig.ident") + RotatedAxis()
plot5 + plot6 + plot7 + plot8 + plot_layout(ncol = 2, guides = 'collect')
rm(list=ls());gc()
```
```{r eval=FALSE}
####----Compute the top relative expression of each gene per cell Use sparse matrix----####
par(mar = c(4, 8, 2, 1))
C <- hswound.com@assays$RNA@counts
C <- Matrix::t(Matrix::t(C)/Matrix::colSums(C)) * 100
most_expressed <- order(apply(C, 1, median), decreasing = T)[20:1]
boxplot(t(as.matrix(C[most_expressed, ])), cex = 0.1, las = 1, xlab = "% total count per cell", 
    col = (scales::hue_pal())(20)[20:1], horizontal = TRUE)
```

## 2.3 filter QC
```{r eval=FALSE, fig.align='center', fig.width=12}
####----Fitler gene expressed in at least 10 cells (>10) -> some uninterested gene hemoglobin genes -> MALAT1 gene -> MT genes----####
length(rownames(hswound.com))
table(hswound.com$orig.ident) #sample number
selected_f <- rownames(hswound.com)[Matrix::rowSums(hswound.com@assays$RNA@counts > 0) > 10]; length(selected_f)
selected_f <- selected_f[-c(grep("^HB[^(PES)]", selected_f))]; length(selected_f)
selected_f <- selected_f[-grep("MALAT1", selected_f)]; length(selected_f)
selected_f <- selected_f[-grep("^MT-", selected_f)]; length(selected_f)
hswound.com #Check first
hswound.com <- subset(hswound.com, features = selected_f)
hswound.com #Check if it works
identical(selected_f, rownames(hswound.com))
hswound.com <- subset(hswound.com, subset = nFeature_RNA > 500 & percent.mt < 20 & nCount_RNA > 1000 & nCount_RNA < 150000)
#after filtering
hswound.com
table(hswound.com$orig.ident)
saveRDS(hswound.com, "../00_Origin_objects/s0_original_seurat_object_noDoublets_afterqc.rds")
```
Cell numbers for each sample after quality control
```{r eval=TRUE, echo=FALSE, warning=FALSE, message=FALSE}
hswound.com <- readRDS("../../00_Origin_objects/s0_original_seurat_object_noDoublets_afterqc.rds")
hswound.com
table(hswound.com$orig.ident)
```

## 2.4 QC plot after filtering
```{r fig.align='center', fig.width=12, fig.height=12}
VlnPlot(
  hswound.com,
  features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.ribo"),
  ncol = 2, pt.size=0)
```

```{r fig.align='center', fig.width=12}
plot1 <- FeatureScatter(hswound.com, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "orig.ident") + RotatedAxis()
plot2 <- FeatureScatter(hswound.com, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by = "orig.ident") + RotatedAxis()
plot3 <- FeatureScatter(hswound.com, feature1 = "nCount_RNA", feature2 = "percent.ribo", group.by = "orig.ident") + RotatedAxis()
plot4 <- FeatureScatter(hswound.com, feature1 = "nCount_RNA", feature2 = "percent.hb", group.by = "orig.ident") + RotatedAxis()
plot1 + plot2 + plot3 + plot4 + plot_layout(ncol = 2, guides = 'collect')
```

```{r fig.align='center', fig.width=12}
plot5 <- FeatureScatter(hswound.com, feature1 = "percent.mt", feature2 = "percent.malat1", group.by = "orig.ident") + RotatedAxis()
plot6 <- FeatureScatter(hswound.com, feature1 = "nCount_RNA", feature2 = "percent.malat1", group.by = "orig.ident") + RotatedAxis()
plot7 <- FeatureScatter(hswound.com, feature1 = "nFeature_RNA", feature2 = "percent.malat1", group.by = "orig.ident") + RotatedAxis()
plot8 <- FeatureScatter(hswound.com, feature1 = "nFeature_RNA", feature2 = "percent.mt", group.by = "orig.ident") + RotatedAxis()
plot5 + plot6 + plot7 + plot8 + plot_layout(ncol = 2, guides = 'collect')
rm(list=ls());gc()
```

#####################################################
# Step 3: Normalization and feature selection
#####################################################
## 3.1 normalization using sctransform for each sample
```{r eval=FALSE}
hswound.list <- SplitObject(hswound.com, split.by = "orig.ident")
####----Do normalization for each dataset----####
####----SCTransform normalization----#### These steps were done in Uppmax
####----SCTransform (default) already normalize the nUMI and nGene count----####
####----Cell cycle: removing all signal associated with cell cycle can negatively impact downstream analysis, particularly in differentiating processes (like murine hematopoiesis), where stem cells are quiescent and differentiated cells are proliferating: regressing out the difference between the G2M and S phase scores. This means that signals separating non-cycling cells and cycling cells will be maintained, but differences in cell cycle phase among proliferating cells (which are often uninteresting) will be regressed out of the data----####
hswound.list <- lapply(X = hswound.list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- CellCycleScoring(object = x, g2m.features = cc.genes$g2m.genes, s.features = cc.genes$s.genes, set.ident = FALSE)
  x$CC.Difference <- x$S.Score - x$G2M.Score
  x <- SCTransform(x, vars.to.regress = c("percent.mt", "CC.Difference"), variable.features.n = 4000, verbose = TRUE, return.only.var.genes = FALSE, method = "glmGamPoi")
  x <- RunPCA(x, npcs = 60, ndims.print = 1:5, nfeatures.print = 5, features = VariableFeatures(object = x))

})
rm(hswound.com);gc() #delete the object to save RAM memory
####----Save the individual seurat object----####
for (i in 1:length(hswound.list)) {
  saveRDS(hswound.list[[i]], file = paste0("s1_cleanSeurat_",names(hswound.list[i]), ".rds"))
}
```

####################################################
# Step 4: Batch-correction
####################################################
## 4.1 Load all the SCT-normalized data
```{r eval=TRUE}
samples <- list.files("../s0_SeuratObject_SplitPerSample/", pattern = ".rds$")
sm_folders <- paste0("../s0_SeuratObject_SplitPerSample/", samples)
samplenames <- gsub("s1_cleanSeurat_", "", samples)
samplenames <- gsub("\\.rds", "", samplenames)
hswound.list <- list()
for (i in seq_along(samples)) {
  tmp_sms <- samplenames[i]
  hswound.list[[tmp_sms]] <- readRDS(sm_folders[i])
}
hswound.list
```
Something need to think about it. (Variable genes across samples/conditions)
```{r eval=FALSE}
####----Check variable genes for each sample----####
vg2 <- lapply(hswound.list, function(x) VariableFeatures(x)[1:3000])
vg.all <- unique(unlist(vg2)) #unique top 3k variable genes among all samples
        
rn <- lapply(hswound.list, rownames) #total number of genes expressed in each sample
nE <- colSums(Reduce(rbind, lapply(rn, function(x) vg.all %in% x))) #Check the overlapping number of variable genes across samples
vg.all <- vg.all[nE == length(hswound.list)] #length(hswound.list) the number of samples
setdiff(features.all, vg.all) #features.all is from SelectIntegrationFeatures function directly (below). PS: no big difference if I didn't do any further filtering even though genes not expressed across all samples
```

## 4.2 Integrate all time points
```{r eval=TRUE, fig.height=5}
####----Integration with SCTransform----####
####----First select top variable genes----####
features.all <- SelectIntegrationFeatures(object.list = hswound.list, nfeatures = 4500) #First select 4.5K genes
rn <- lapply(hswound.list, rownames) 
nE <- colSums(Reduce(rbind, lapply(rn, function(x) features.all %in% x))) 
features.all <- features.all[nE == length(hswound.list)][1:4000] #Keep only 4K variable genes expressed in all samples

eachvg <- list(
  PWH26D0=VariableFeatures(hswound.list$PWH26D0), 
  PWH26D1=VariableFeatures(hswound.list$PWH26D1),
  PWH26D30=VariableFeatures(hswound.list$PWH26D30),
  PWH26D7=VariableFeatures(hswound.list$PWH26D7),
  PWH27D0=VariableFeatures(hswound.list$PWH27D0),
  PWH27D1=VariableFeatures(hswound.list$PWH27D1),
  PWH27D30=VariableFeatures(hswound.list$PWH27D30),
  PWH27D7=VariableFeatures(hswound.list$PWH27D7),
  PWH28D0=VariableFeatures(hswound.list$PWH28D0),
  PWH28D1=VariableFeatures(hswound.list$PWH28D1),
  PWH28D30=VariableFeatures(hswound.list$PWH28D30),
  PWH28D7=VariableFeatures(hswound.list$PWH28D7),
  Allvg=features.all)
source("../Functions/overlap_phyper2.R")
overlap_phyper2(eachvg, eachvg)
```
```{r eval=FALSE, fig.height=6, fig.width=8}
####----Prepare the integration----####
hswound.list <- PrepSCTIntegration(object.list = hswound.list, anchor.features = features.all)

hswound.combined.sct <- merge(hswound.list[[1]], 
                         y = hswound.list[2:length(hswound.list)],
                         project = "humanwound",
                         merge.data = TRUE)
hswound.combined.sct
```
```{r echo=FALSE,warning=FALSE,message=FALSE}
rm(list = ls());gc()
hswound.combined.sct <- readRDS("s1_cleanSeurat_harmony_allSamples_clusters.rds")
```
```{r eval=FALSE, fig.height=6, fig.width=8}
VariableFeatures(hswound.combined.sct) <- features.all
hswound.combined.sct <- RunPCA(object = hswound.combined.sct, assay = "SCT", npcs = 60, features = VariableFeatures(hswound.combined.sct))
```
```{r eval=TRUE, fig.height=6, fig.width=8}
table(hswound.combined.sct$orig.ident)
ElbowPlot(hswound.combined.sct, ndims = 60, reduction = "pca")
```
```{r eval=TRUE, fig.height=15}
####----Check the genes contributing to top PCs----####
#If there is any gene related to the ribosomal genes (^RP[S|L]),
VizDimLoadings(hswound.combined.sct, dims = 1:12, reduction = "pca")
```
```{r eval=FALSE}
hswound.combined.sct <- RunHarmony(object = hswound.combined.sct,
                                    assay.use = "SCT",
                                    reduction = "pca",
                                    dims.use = 1:40,
                                    group.by.vars = "orig.ident",
                                    plot_convergence = TRUE)

#rn pca, umap, clustering
hswound.combined.sct <- RunUMAP(object = hswound.combined.sct, assay = "SCT", reduction = "harmony", dims = 1:40)
hswound.combined.sct <- FindNeighbors(object = hswound.combined.sct, assay = "SCT", reduction = "harmony", dims = 1:40, k.param = 40)
hswound.combined.sct <- FindClusters(object = hswound.combined.sct, resolution = 0.5)
hswound.combined.sct <- FindClusters(object = hswound.combined.sct, resolution = 0.8)
hswound.combined.sct$orig.ident <- factor(x = hswound.combined.sct$orig.ident, levels =
                                            c("PWH26D0", "PWH26D1", "PWH26D7","PWH26D30",
                                              "PWH27D0", "PWH27D1", "PWH27D7","PWH27D30",
                                              "PWH28D0", "PWH28D1", "PWH28D7","PWH28D30"))
hswound.combined.sct$Condition <- factor(x = hswound.combined.sct$Condition, levels =
                                            c("Skin", "Wound1", "Wound7","Wound30"))
saveRDS(hswound.combined.sct, file = "s1_cleanSeurat_harmony_allSamples_clusters.rds")
```
## 4.3 UMAP plots of Clusters, Sampled, Group types, Gender
```{r}
DimPlot(hswound.combined.sct, reduction = "umap", group.by='seurat_clusters', label=TRUE) +
  ggtitle('Cluster')
```
```{r}
DimPlot(hswound.combined.sct, reduction = "umap", group.by='orig.ident', label=FALSE) +
  ggtitle('Samples')
```
```{r fig.align='center', fig.height=10, fig.width=12}
DimPlot(hswound.combined.sct, reduction = "umap", group.by = "seurat_clusters", pt.size = .001, split.by = 'orig.ident', ncol = 4, label = TRUE) + NoLegend()
```
```{r fig.height=6}
DimPlot(hswound.combined.sct, reduction = "umap", group.by='Condition', label=FALSE) +
  ggtitle('Groups')
DimPlot(hswound.combined.sct, reduction = "umap", group.by = "seurat_clusters", pt.size = .001, split.by = 'Condition', ncol = 4, label = FALSE) + NoLegend()
```
```{r}
DimPlot(hswound.combined.sct, reduction = "umap", group.by='Age', label=FALSE) +
  ggtitle('Age')
```
```{r}
DimPlot(hswound.combined.sct, reduction = "umap", group.by='Gender', label=FALSE) +
  ggtitle('Gender')
```
```{r fig.height=6}
# Explore whether clusters segregate by cell cycle phase
DimPlot(hswound.combined.sct,
        label = TRUE,
        split.by = "Phase")  + NoLegend()

(DimPlot(hswound.combined.sct, reduction = "umap", group.by='seurat_clusters', label=TRUE) + ggtitle('Clusters') + NoAxes() + NoLegend()) +
  (FeaturePlot(hswound.combined.sct, features = c("MKI67"), reduction = "umap", cols = c("gray","red")) + NoLegend() + NoAxes())
```
```{r}
# Determine metrics to plot present in hswound.combined.sct@meta.data
metrics <-  c("nCount_RNA", "nFeature_RNA", "S.Score", "G2M.Score", "percent.mt")

FeaturePlot(hswound.combined.sct,
            reduction = "umap",
            features = metrics,
            pt.size = 0.001,
            label = FALSE)
```

## 4.4 stacked barplot showing the proportion of cells from each batch in each cluster
```{r}
####----Please pay attentions to the original numbers of cells per sample----####
clusters <- unique(hswound.combined.sct@meta.data[["seurat_clusters"]])
clusters <- clusters[order(clusters)]
df <- data.frame()
for(i in 1:length(clusters)){
  SmCell_sum <- table(hswound.combined.sct$orig.ident)
  tmp.df1 <- hswound.combined.sct@meta.data %>% subset(seurat_clusters == clusters[i]) %>% select(orig.ident) %>% table()
  if(length(tmp.df1) == 12){
    #First normalize the total cell counts per sample
    cur_df <- as.data.frame(tmp.df1 / SmCell_sum)
    colnames(cur_df) <- c("Sample", "Freq")
    #Calculate the normalized proportion 
    cur_df$Freq <- cur_df$Freq * 1/(sum(cur_df$Freq))
    cur_df$Cluster <- clusters[i]
    df <- rbind(df, cur_df)
  } else {
    #print(i);print(tmp.df1)
    #only include the matched samples
    match.sample <- SmCell_sum[names(SmCell_sum) %in% names(tmp.df1)]
    #print(match.sample)
    #First normalize the total cell counts per sample
    cur_df <- as.data.frame(tmp.df1 / match.sample)
    colnames(cur_df) <- c("Sample", "Freq")
    #Calculate the normalized proportion 
    cur_df$Freq <- cur_df$Freq * 1/(sum(cur_df$Freq))
    cur_df$Cluster <- clusters[i]
    df <- rbind(df, cur_df)
  }
}

##Plot for each sample
ggplot(df, aes(x = Cluster, y = Freq, fill = Sample)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_brewer(palette = "Paired") + # Paired or Set3 maximum variable = 12
  xlab('Clsuter') +
  scale_y_continuous(breaks = seq(0, 1, 0.1),
                     expand = c(0, 0),
                     name = 'Percentage') +
  theme_bw() +
  theme(
    panel.grid.major.x = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_blank(),
    #panel.border = element_rect(size = 0.35),
    axis.ticks = element_blank()#,
    #axis.text.x = element_text(size = 5)
  ) +
  coord_flip()

##Plot for each group
df.group <- df %>% mutate(Sample = gsub("PWH..", "", Sample)) %>% group_by(Cluster, Sample) %>% 
  summarise(Freq = sum(Freq))
df.group$Sample <- factor(df.group$Sample, levels = c("D0", "D1", "D7", "D30"))
ggplot(df.group, aes(x = Cluster, y = Freq, fill = Sample)) +
  geom_bar(position = "stack", stat = "identity") +
  scale_fill_brewer(palette = "Set1") + # Paired or Set3 maximum variable = 12
  xlab('Clsuter') +
  scale_y_continuous(breaks = seq(0, 1, 0.1),
                     expand = c(0, 0),
                     name = 'Percentage') +
  theme_bw() +
  theme(
    panel.grid.major.x = element_blank(),
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_blank(),
    #panel.border = element_rect(size = 0.35),
    axis.ticks = element_blank()#,
    #axis.text.x = element_text(size = 5)
  ) +
  coord_flip()
```

## 4.5 Clusters determined by resolutions
```{r eval=TRUE, fig.width=12, fig.height=22, warning=FALSE, message=FALSE}
# Clustering with louvain (algorithm 1) using different resolutions
for (res in c(0.2, 0.4, 0.5, 0.6, 0.8, 1.0, 1.2, 1.5, 2)) {
    hswound.combined.sct <- FindClusters(hswound.combined.sct, resolution = res, verbose = FALSE)
}

plot_grid(ncol = 2, 
          DimPlot(hswound.combined.sct, reduction = "umap", group.by = "SCT_snn_res.0.2") +
            ggtitle("louvain_0.2"), 
          DimPlot(hswound.combined.sct, reduction = "umap", group.by = "SCT_snn_res.0.4") + 
            ggtitle("louvain_0.4"), 
          DimPlot(hswound.combined.sct, reduction = "umap", group.by = "SCT_snn_res.0.5") + 
            ggtitle("louvain_0.5"),
          DimPlot(hswound.combined.sct, reduction = "umap", group.by = "SCT_snn_res.0.6") +
            ggtitle("louvain_0.6"), 
          DimPlot(hswound.combined.sct, reduction = "umap", group.by = "SCT_snn_res.0.8") + 
            ggtitle("louvain_0.8"), 
          DimPlot(hswound.combined.sct, reduction = "umap", group.by = "SCT_snn_res.1") + 
            ggtitle("louvain_1"),
          DimPlot(hswound.combined.sct, reduction = "umap", group.by = "SCT_snn_res.1.2") +
            ggtitle("louvain_1.2"), 
          DimPlot(hswound.combined.sct, reduction = "umap", group.by = "SCT_snn_res.1.5") + 
            ggtitle("louvain_1.5"), 
          DimPlot(hswound.combined.sct, reduction = "umap", group.by = "SCT_snn_res.2") + 
            ggtitle("louvain_2"))
```

## 4.6 Cluster tree plot
```{r eval=TRUE, fig.width=12, fig.height=18}
# install.packages('clustree')
suppressPackageStartupMessages(library(clustree))

clustree(hswound.combined.sct@meta.data, prefix = "SCT_snn_res.")
```
```{r eval=TRUE, echo=FALSE, warning=FALSE, message=FALSE}
rm(list = ls());gc(verbose = FALSE)
```

################################################
# Step 5. DE analysis using original RNA assay
################################################
## 5.1 FindAllMarkers
```{r eval=FALSE}
####----Change to RNA assay to do the DE analysis----####
DefaultAssay(hswound.combined.sct) <- "RNA" #DE for all genes instead of only using variable gene
hswound.combined.sct

hswound.combined.sct <- NormalizeData(hswound.combined.sct, verbose = TRUE,  normalization.method = "LogNormalize", scale.factor = 10000)

####----Calculate the scale data for following heatmap visualization of marker genes----####
hswound.combined.sct <- ScaleData(object = hswound.combined.sct, features = rownames(hswound.combined.sct), vars.to.regress = c("percent.mt", "CC.Difference", "nCount_RNA", "nFeature_RNA"))
#save data
saveRDS(hswound.combined.sct, "s2_cleanSeurat_harmony_allSamples_clusters_normGenes.rds") #save normalized gene expressions

Idents(hswound.combined.sct) <- "SCT_snn_res.0.8" 
table(hswound.combined.sct$SCT_snn_res.0.8)
####----FindAllMarkers for all clusters----####
markers <- FindAllMarkers(
  hswound.combined.sct,
  only.pos = TRUE,
  min.pct = 0.25,  #min.pct = 0.5, logfc.threshold = 0.5,
  logfc.threshold = 0.25,
  test.use = "MAST") #MAST has good FDR control and is faster than DESeq2, default test is wilcox

####----add the gene description to the cluster marker genes----####
require(biomaRt) #listMarts() : Check the version and dataset
hs.ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl") #listAttributes(mart = hs.ensembl)
annot <- getBM(attributes = c("external_gene_name", "entrezgene_id", "gene_biotype","description"), mart=hs.ensembl) %>% 
  distinct(external_gene_name, .keep_all = TRUE) %>% tidyr::separate(description, c("Gene_anno", "Other"), sep="\\[Source") %>% 
  dplyr::select(-5)

annot <- readRDS("../Functions/humanAnnotation.rds")
####----add the info and reorder the columns----####
markers <- markers %>% left_join(., annot, by=c("gene" = "external_gene_name")) %>% 
  dplyr::select(6, 7, 2:4, 1, 5, everything()) %>% arrange(cluster, desc(avg_log2FC))
data.table::fwrite(markers, file="s2_cleanSeurat_harmony_allSamples_markerGene_ori.txt", sep="\t")
```
```{r}
hswound.combined.sct <- readRDS(file = "s2_cleanSeurat_harmony_allSamples_clusters_normGenes.rds") #reload the clustering results with normalized expressions for further visualization
hswound.combined.sct
variable.genes <- VariableFeatures(hswound.combined.sct, assay="SCT") %>% as.data.frame() %>% rename("gene" = ".") %>% 
  mutate(VariableGene = "Yes")
markers <- data.table::fread("s2_cleanSeurat_harmony_allSamples_markerGene_ori.txt") #marker genes
####----Add the information about if the gene is the variable gene----####
markers <- markers %>% left_join(., variable.genes, by=c("gene" = "gene"))
```

## 5.2 Draw marker plots (Heatmap and Violin plot)
```{r fig.height=20}
top5 <- markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC)
####----Heatmap plot----####
DoHeatmap(hswound.combined.sct, features = as.character(unique(top5$gene)), group.by='SCT_snn_res.0.8', label=TRUE, size = 2, angle = 30, disp.min = -2.5)
```

```{r fig.height=42}
####----Violin plot----####
top1 <- markers %>% group_by(cluster) %>% top_n(n = 1, wt = avg_log2FC)
VlnPlot(hswound.combined.sct, features = top1$gene, group.by='SCT_snn_res.0.8', ncol = 2, pt.size = 0) 
```

## 5.3 Check known marker genes
```{r eval=TRUE, fig.height=12, fig.width=12}
# set up list of canonical cell type markers
canonical_markers <- list(
  'Keratinocytes' = c('KRT5', 'KRT14', 'KRT1', 'KRT10'),
  'Fibroblasts' = c('COL1A1', 'COL1A2', 'DCN'),
  'Melanocyte' =	c('PMEL',	'MLANA'),
  'Endothelial-cells' = c('PECAM1', 'VWF'),
  'T-cells' = c('CD52', 'CD3D'),
  'Dendritic-cells' = c('CD74')
)
plot_list <- FeaturePlot(
  hswound.combined.sct,
  features=unlist(canonical_markers),
  combine=FALSE, cols = c("gray","red")
)

# apply theme to each feature plot
for(i in 1:length(plot_list)){
  plot_list[[i]] <- plot_list[[i]] + NoLegend() + NoAxes()
}

cowplot::plot_grid(plotlist = plot_list, ncol = 4)
rm(plot_list);gc(verbose = FALSE)
```

## 5.4 Annotate clusters by using R packages
```{r eval=TRUE}
library(SingleR)
library(celldex)
hpca.se <- HumanPrimaryCellAtlasData()
hpca.se$label.main <- str_replace_all(hpca.se$label.main, 
                                      c("NK_cell" = "NK cells", "B_cell" = "B cells", 
                                        "DC" = "Dendritic cells","HSC_-G-CSF" = "HSC G-CSF",
                                        "Monocyte" = "Monocytes", "Astrocyte" = "Astrocytes",
                                        "Erythroblast" = "Erythroblasts", "Macrophage" = "Macrophages",
                                        "BM" = "BMs", "MSC"="MSCs",
                                        "CMP" = "CMPs", "GMP" = "GMPs", 
                                        "MEP"="MEPs", "Myelocyte"="Myelocytes",
                                        "Neuroepithelial_cell"="Neuroepithelial cells", "T_cells" = "T cells", 
                                        "iPS_cells" = "iPS cells", "Endothelial_cells" = "Endothelial cells", 
                                        "Tissue_stem_cells" = "Tissue stem cells", "Embryonic_stem_cells" = "Embryonic stem cells",
                                        "Smooth_muscle_cells" = "Smooth muscle cells", "Epithelial_cells" = "Epithelial cells"))
hpca.se$label.main[hpca.se$label.main == "HSC_CD34+"] <-"HSC CD34"
hpca.se$label.main[hpca.se$label.main == "Pro-B cells_CD34+"] <-"Pro-B cells CD34 Pos"
hpca.se$label.main[hpca.se$label.main == "Pre-B cells_CD34-"] <-"Pre-B cells CD34 Neg"

####----Check the unique cell types----####
hpca.se$label.main %>% unique()

#Get the count data from Seurat Object
hswound.singleR <- GetAssayData(hswound.combined.sct)
##First check the main cell type
Idents(hswound.combined.sct) <- "SCT_snn_res.0.8" #Choose different resolutions to annotate
pred.hesc <- SingleR(test = hswound.singleR, ref = hpca.se, assay.type.test=1, labels = hpca.se$label.main, clusters = Idents(hswound.combined.sct))
##Plot of main cell type
plotScoreHeatmap(pred.hesc)

celltype.main <- as.data.frame(pred.hesc) %>% tibble::rownames_to_column(var = "Cluster")
celltype.main.score <- as.data.frame(pred.hesc) %>% t() %>% as.data.frame() %>% tibble::rownames_to_column(var = "Cluster") %>%  mutate(SingleR = "MainCellType")

celltype.main$Cluster <- parse_integer(celltype.main$Cluster)
markers <- markers %>% left_join(., celltype.main[, c(1,38)], by=c("cluster" = "Cluster")) #Choose the first.label

####----Check the unique fine cell types----####
hpca.se$label.fine %>% unique() #sub cell types [which represents the highest resolution]
##Second check the fine cell type
pred.hesc.fine <- SingleR(test = hswound.singleR, ref = hpca.se, assay.type.test=1, labels = hpca.se$label.fine, clusters = Idents(hswound.combined.sct))
##Plot of fine cell type
plotScoreHeatmap(pred.hesc.fine)

celltype.fine <- as.data.frame(pred.hesc.fine) %>% tibble::rownames_to_column(var = "Cluster")
celltype.fine.score <- as.data.frame(pred.hesc.fine) %>% t() %>% as.data.frame() %>% tibble::rownames_to_column(var = "Cluster") %>% mutate(SingleR = "FineCellType")

celltype.fine$Cluster <- parse_integer(celltype.fine$Cluster)
markers <- markers %>% left_join(., celltype.fine[, c(1,159)], by=c("cluster" = "Cluster")) #Choose the first.label

colnames(markers)[12:13] <- c("SingleR_main", "SingleR_fine")
data.table::fwrite(markers, "s2_cleanSeurat_harmony_allSamples_markerGene.txt", sep = "\t")

celltype.scores <- rbind(celltype.main.score, celltype.fine.score) %>% dplyr::select(29, everything())
data.table::fwrite(celltype.scores, file = "s2_cleanSeurat_harmony_allSamples_markerGene_singleR_scores.txt", sep = "\t")
```

# Sessioninfo
```{r}
sessionInfo()
```
