---
title: "Cross-species analysis"
author: "Zhuang Liu"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true #add a table of contents (TOC)
    toc_depth: 3 #headers' level in the toc
    toc_float: true #toc visible even when scrolled
    theme: lumen
    highlight: tango
    code_folding: show
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align='center', fig.width=12, fig.height = 6, warning=FALSE, message=FALSE)
```

```{r warning=FALSE, message=FALSE}
library(Seurat)
library(SeuratObject)
library(SeuratDisk)
library(tidyverse)
library(patchwork)
library(harmony)
#library(scCustomize)
```

# 3. Integrating human and mouse (in-house, CSCpaper) wound scRNA-seq
## 3.1 Converting mouse genes to human orthologous genes
```{r eval=FALSE}
library(Matrix)
#######################
# 1.Mouse wound scRNA-seq

# 2.Load the CSC wound healing (From China national database)
ms_inte <- readRDS("/Users/zhuliu/Desktop/ZhuangLiu/00_Projects/s2_hsms_wound_diff/03-RawData/sc05_CSC_GSACRA010641_wholetissue.rds")
DefaultAssay(ms_inte) <- "RNA"
ms_inte <- DietSeurat(ms_inte, assays = "RNA")

# Remove unneeded columns of metadata
ms_inte$Condition <- gsub("skin_", "", ms_inte$Condition)
table(ms_inte$Condition)
ms_inte$Project <- "cscChina"

rownames(ms_inte)[grep("Mmp", rownames(ms_inte))]

# load the conversion of human_mouse genes
hs_mm <- data.table::fread("/Users/zhuliu/Desktop/sc st/Versions_SC_ST_Project/zCell_multiJournal_revision/01-Revision_scripts_figs/Dtmp_figures/00-gene_conversion_hs_ms/Private_human_mouse_GeneSymbol_conversion.txt")

exp_mtx <- as.matrix(ms_inte@assays$RNA@counts);dim(exp_mtx)
v2genes <- data.frame(ms_gene = rownames(exp_mtx)) %>% left_join(., hs_mm, by=c("ms_gene"))

table(is.na(v2genes$hs_gene))
##lots of NAs for which there are no human genes matching
## Remove NAs
v2genes <- v2genes[!is.na(v2genes$hs_gene),,F]
## Filter the expression matrix for genes which a mouse counterpart is available
exp_mtx <- exp_mtx[v2genes$ms_gene,]
## Now change the rownames of the matrix to the mouse gene names
rownames(exp_mtx) <- v2genes$hs_gene
#check the duplicated gene names
duplicated(rownames(exp_mtx)) %>% table()
exp_mtx <- exp_mtx[!duplicated(rownames(exp_mtx)),]
dim(exp_mtx);identical(colnames(exp_mtx), colnames(ms_inte))
## Create the seurat object with mouse genes.
ms_inte <- CreateSeuratObject(counts = exp_mtx, meta.data = ms_inte@meta.data)
rownames(ms_inte)[grep("KRT", rownames(ms_inte))]
rownames(ms_inte)[grep("MMP", rownames(ms_inte))]
rownames(ms_inte)[grep("S100", rownames(ms_inte))]
rownames(ms_inte)[grep("SERPINB", rownames(ms_inte))]

saveRDS(ms_inte, file = "cscChina_mouse_wound.rds")
```

## 3.2 Merge three datasets together
```{r}
# Human skin wound healing 
hs_seu <- readRDS("all_human_wound_metadata.rds")
hs_seu
mt_hs <- hs_seu@meta.data %>% rownames_to_column(var = "barcode")
mt_hs$Project <- "HumanWound"
mt_hs$Sample <- mt_hs$orig.ident
mt_hs <- mt_hs %>% column_to_rownames(var = "barcode") %>% select(1,7,4:6)
hs_seu@meta.data <- mt_hs

# Mouse skin wound healing
mm_seu <- readRDS("cscChina_mouse_wound.rds")
mm_seu 
mt_mm <- mm_seu@meta.data %>% rownames_to_column(var = "barcode")
mt_mm$Sample <- mt_mm$orig.ident
mt_mm <- mt_mm %>% select(1,2,9,6,7,8) 
mt_mm <- mt_mm %>% column_to_rownames(var = "barcode")
colnames(mt_mm) <- colnames(mt_hs)
mm_seu@meta.data <- mt_mm

# Merge data
# check the overlapped genes and keep the overlapped genes
overgene <- intersect(rownames(hs_seu), rownames(mm_seu))
length(overgene)
mm_seu <- mm_seu[overgene, ]
hs_seu <- hs_seu[overgene, ]

# integrate based on orig.ident names
mm_seu$Species="Mouse"
hs_seu$Species="Human"

all_seu <- merge(hs_seu,  y = c(mm_seu),
                project = "wounds")

table(all_seu$Project)
rownames(all_seu)[grep("KRT", rownames(all_seu))]
rownames(all_seu)[grep("MMP", rownames(all_seu))]
rownames(all_seu)[grep("S100", rownames(all_seu))]
rownames(all_seu)[grep("SERPINB", rownames(all_seu))]
```

## 3.3 noSampling (First) and normalization, CCA clustering
```{r}
# sampling the cell numbers 
table(all_seu$Condition, all_seu$Project)
#lowNum <- 594
#metadata <- all_seu@meta.data %>% rownames_to_column(var = "barcode")
#to_keep <- metadata %>% group_by(Condition, Project) %>% sample_n(size = lowNum, replace = FALSE) %>%
#    pull(barcode)
#all_seu <- subset(all_seu, cells = to_keep)
#table(all_seu$Condition,all_seu$Project)

# split the data
alldata = SplitObject(all_seu, split.by = "orig.ident")
rm(all_seu);gc()
# normalize and identify variable features for each dataset independently
alldata <- lapply(X = alldata, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 4000)
})

# Run CCA using defaults
features <- SelectIntegrationFeatures(object.list = alldata, nfeatures = 3000)
anno <- readRDS("./00-gene_conversion_hs_ms/humanAnnotation.rds")
features <- features %>% as.data.frame() %>% setNames("gene") %>% left_join(., anno[, c(1,3)], by=c("gene"="external_gene_name")) %>% 
  filter(gene_biotype == "protein_coding" | gene_biotype == "lncRNA") %>% distinct(gene) %>% pull(gene)

rn <- lapply(alldata, rownames) 
nE <- colSums(Reduce(rbind, lapply(rn, function(x) features %in% x))) 
features <- features[nE == length(alldata)] #Keep only variable genes expressed in all samples
length(features)

# Integrate alldata
alldata.anchors <- FindIntegrationAnchors(object.list = alldata, anchor.features = features)
# this command creates an 'integrated' data assay
inteData <- IntegrateData(anchorset = alldata.anchors)

# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the 'RNA' assay
DefaultAssay(inteData) <- "integrated"
VariableFeatures(inteData) <- features

# Run the standard workflow for visualization and clustering
inteData <- ScaleData(inteData)
inteData <- RunPCA(inteData, npcs = 30)

inteData <- RunUMAP(inteData, dims = 1:30, assay = "integrated", reduction = "pca", n.neighbors = 40,
                    min.dist = 0.5, n.epochs = 500, spread = 1, learning.rate = 1)

DimPlot(inteData, group.by = "orig.ident") + NoAxes()
DimPlot(inteData, group.by = "Condition", split.by = "Project")
DimPlot(inteData, group.by = "CellType", split.by = "Project", label = T)
#DimPlot_scCustom(inteData, group.by = "CellType", split.by = "Condition", num_columns = 5, label = T, pt.size = 0.2)
FeaturePlot(inteData, features = c("S100A8"))

inteData = FindNeighbors(inteData, dims = 1:30, reduction = "pca", k.param = 40)
inteData = FindClusters(inteData, resolution = 0.3)
inteData = FindClusters(inteData, resolution = 1)
inteData = FindClusters(inteData, resolution = 0.5)
inteData = FindClusters(inteData, resolution = 0.8)

DefaultAssay(inteData) <- "RNA"
inteData <- NormalizeData(inteData)
```

## 3.4 CellType plotting
```{r fig.width=12, fig.height=8}
DimPlot(inteData, group.by = "Species", cols = c("#d8b365", "#5ab4ac")) + ggtitle("Cross species integration")
DimPlot(inteData, group.by = "Species", split.by = "Species", cols = c("#d8b365", "#5ab4ac")) + ggtitle("Cross species integration")

DimPlot(inteData, group.by = "CellType", label = T, split.by = "Species") + ggtitle("")
DimPlot(inteData, group.by = "CellType", label = T, split.by = "Condition", ncol = 4) + ggtitle("") + NoLegend()
DimPlot(inteData, group.by = "integrated_snn_res.0.3", label = T, split.by = "Condition")
DimPlot(inteData, group.by = "integrated_snn_res.0.5", label = T, split.by = "Condition")
DimPlot(inteData, group.by = "integrated_snn_res.0.8", label = T, split.by = "Condition")
DimPlot(inteData, group.by = "integrated_snn_res.1", label = T, split.by = "Condition")
```


# 4. Data exporting
```{r}
saveRDS(inteData, "all_human_mouse_integration_cscChina.rds")

# export the reduced results for quickly plotting
inteData@reductions$pca@assay.used <- "RNA"
inteData@reductions$umap@assay.used <- "RNA"

inteData <- DietSeurat(inteData, assays = "RNA", dimreducs = c("umap"))
saveRDS(inteData, "all_human_mouse_integration_cscChina_reduced.rds")
```


# SessionInfo
```{r}
sessionInfo()
```

