---
title: "Cross-species analysis"
author: "Zhuang Liu"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true #add a table of contents (TOC)
    toc_depth: 3 #headers' level in the toc
    toc_float: true #toc visible even when scrolled
    theme: lumen
    highlight: tango
    code_folding: show
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align='center', fig.width=12, fig.height = 6, warning=FALSE, message=FALSE)
```

```{r warning=FALSE, message=FALSE}
library(Seurat)
library(SeuratObject)
library(SeuratDisk)
library(tidyverse)
library(patchwork)
library(harmony)
library(scCustomize)
library(scRNAtoolVis)
```

# 3. Integrating human and mouse (in-house, CSCpaper) wound scRNA-seq
## 3.1 Converting mouse genes to human orthologous genes
```{r eval=FALSE}
library(Matrix)
#######################
# 1.Mouse wound scRNA-seq

# 2.Load the CSC wound healing (From China national database)
ms_inte <- readRDS("/Users/zhuliu/Desktop/ZhuangLiu/00_Projects/s2_hsms_wound_diff/03-RawData/sc05_CSC_GSACRA010641_wholetissue.rds")
DefaultAssay(ms_inte) <- "RNA"
ms_inte <- DietSeurat(ms_inte, assays = "RNA")

# Remove unneeded columns of metadata
ms_inte$Condition <- gsub("skin_", "", ms_inte$Condition)
table(ms_inte$Condition)
ms_inte$Project <- "cscChina"

rownames(ms_inte)[grep("Mmp", rownames(ms_inte))]

# load the conversion of human_mouse genes
hs_mm <- data.table::fread("/Users/zhuliu/Desktop/sc st/Versions_SC_ST_Project/zCell_multiJournal_revision/01-Revision_scripts_figs/Dtmp_figures/00-gene_conversion_hs_ms/Private_human_mouse_GeneSymbol_conversion.txt")

exp_mtx <- as.matrix(ms_inte@assays$RNA@counts);dim(exp_mtx)
v2genes <- data.frame(ms_gene = rownames(exp_mtx)) %>% left_join(., hs_mm, by=c("ms_gene"))

table(is.na(v2genes$hs_gene))
##lots of NAs for which there are no human genes matching
## Remove NAs
v2genes <- v2genes[!is.na(v2genes$hs_gene),,F]
## Filter the expression matrix for genes which a mouse counterpart is available
exp_mtx <- exp_mtx[v2genes$ms_gene,]
## Now change the rownames of the matrix to the mouse gene names
rownames(exp_mtx) <- v2genes$hs_gene
#check the duplicated gene names
duplicated(rownames(exp_mtx)) %>% table()
exp_mtx <- exp_mtx[!duplicated(rownames(exp_mtx)),]
dim(exp_mtx);identical(colnames(exp_mtx), colnames(ms_inte))
## Create the seurat object with mouse genes.
ms_inte <- CreateSeuratObject(counts = exp_mtx, meta.data = ms_inte@meta.data)
rownames(ms_inte)[grep("KRT", rownames(ms_inte))]
rownames(ms_inte)[grep("MMP", rownames(ms_inte))]
rownames(ms_inte)[grep("S100", rownames(ms_inte))]
rownames(ms_inte)[grep("SERPINB", rownames(ms_inte))]

saveRDS(ms_inte, file = "cscChina_mouse_wound.rds")
```

## 3.2 Merge three datasets together
```{r eval=FALSE}
# Human skin wound healing 
hs_seu <- readRDS("all_human_wound_metadata.rds")
hs_seu
mt_hs <- hs_seu@meta.data %>% rownames_to_column(var = "barcode")
mt_hs$Project <- "HumanWound"
mt_hs$Sample <- mt_hs$orig.ident
mt_hs <- mt_hs %>% column_to_rownames(var = "barcode") %>% select(1,7,4:6)
hs_seu@meta.data <- mt_hs

# Mouse skin wound healing
mm_seu <- readRDS("cscChina_mouse_wound.rds")
mm_seu 
mt_mm <- mm_seu@meta.data %>% rownames_to_column(var = "barcode")
mt_mm$Sample <- mt_mm$orig.ident
mt_mm <- mt_mm %>% select(1,2,9,6,7,8) 
mt_mm <- mt_mm %>% column_to_rownames(var = "barcode")
colnames(mt_mm) <- colnames(mt_hs)
mm_seu@meta.data <- mt_mm

# Merge data
# check the overlapped genes and keep the overlapped genes
overgene <- intersect(rownames(hs_seu), rownames(mm_seu))
length(overgene)
mm_seu <- mm_seu[overgene, ]
hs_seu <- hs_seu[overgene, ]

# integrate based on orig.ident names
mm_seu$Species="Mouse"
hs_seu$Species="Human"

all_seu <- merge(hs_seu,  y = c(mm_seu),
                project = "wounds")

table(all_seu$Project)
rownames(all_seu)[grep("KRT", rownames(all_seu))]
rownames(all_seu)[grep("MMP", rownames(all_seu))]
rownames(all_seu)[grep("S100", rownames(all_seu))]
rownames(all_seu)[grep("SERPINB", rownames(all_seu))]
```

## 3.3 Sampling condition and normalization, CCA clustering
```{r eval=FALSE}
# sampling the cell numbers 
table(all_seu$Condition, all_seu$Project)
table(all_seu$orig.ident, all_seu$Project)
lowNum <- 5631
metadata <- all_seu@meta.data %>% rownames_to_column(var = "barcode")
to_keep <- metadata %>% group_by(Condition, Project) %>% sample_n(size = lowNum, replace = FALSE) %>%
pull(barcode)
all_seu <- subset(all_seu, cells = to_keep)
table(all_seu$Condition,all_seu$Project)

# split the data
alldata = SplitObject(all_seu, split.by = "orig.ident")
rm(all_seu);gc()
# normalize and identify variable features for each dataset independently
alldata <- lapply(X = alldata, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 4000)
})

# Run CCA using defaults
features <- SelectIntegrationFeatures(object.list = alldata, nfeatures = 3000)
anno <- readRDS("./00-gene_conversion_hs_ms/humanAnnotation.rds")
features <- features %>% as.data.frame() %>% setNames("gene") %>% left_join(., anno[, c(1,3)], by=c("gene"="external_gene_name")) %>% 
  filter(gene_biotype == "protein_coding" | gene_biotype == "lncRNA") %>% distinct(gene) %>% pull(gene)

rn <- lapply(alldata, rownames) 
nE <- colSums(Reduce(rbind, lapply(rn, function(x) features %in% x))) 
features <- features[nE == length(alldata)] #Keep only variable genes expressed in all samples
length(features)

# Integrate alldata
alldata.anchors <- FindIntegrationAnchors(object.list = alldata, anchor.features = features)
# this command creates an 'integrated' data assay
inteData <- IntegrateData(anchorset = alldata.anchors)

# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the 'RNA' assay
DefaultAssay(inteData) <- "integrated"
VariableFeatures(inteData) <- features

# Run the standard workflow for visualization and clustering
inteData <- ScaleData(inteData)
inteData <- RunPCA(inteData, npcs = 30)

inteData <- RunUMAP(inteData, dims = 1:30, assay = "integrated", reduction = "pca", n.neighbors = 40,
                    min.dist = 0.5, n.epochs = 500, spread = 1, learning.rate = 1)

DimPlot(inteData, group.by = "orig.ident") + NoAxes()
DimPlot(inteData, group.by = "Condition", split.by = "Project")
DimPlot(inteData, group.by = "CellType", split.by = "Project", label = T)
#DimPlot_scCustom(inteData, group.by = "CellType", split.by = "Condition", num_columns = 5, label = T, pt.size = 0.2)
FeaturePlot(inteData, features = c("S100A8"))

inteData = FindNeighbors(inteData, dims = 1:30, reduction = "pca", k.param = 40)
inteData = FindClusters(inteData, resolution = 0.3)
inteData = FindClusters(inteData, resolution = 1)
inteData = FindClusters(inteData, resolution = 0.5)
inteData = FindClusters(inteData, resolution = 0.8)

DefaultAssay(inteData) <- "RNA"
inteData <- NormalizeData(inteData)
```

## 3.4 CellType plotting
```{r eval=FALSE,fig.width=12, fig.height=8}
DimPlot(inteData, group.by = "Species", cols = c("#d8b365", "#5ab4ac")) + ggtitle("Cross species integration")
DimPlot(inteData, group.by = "Species", split.by = "Species", cols = c("#d8b365", "#5ab4ac")) + ggtitle("Cross species integration")

DimPlot(inteData, group.by = "CellType", label = T, split.by = "Species") + ggtitle("")
DimPlot(inteData, group.by = "CellType", label = T, split.by = "Condition", ncol = 4) + ggtitle("") + NoLegend()
DimPlot(inteData, group.by = "integrated_snn_res.0.3", label = T, split.by = "Condition")
DimPlot(inteData, group.by = "integrated_snn_res.0.5", label = T, split.by = "Condition")
DimPlot(inteData, group.by = "integrated_snn_res.0.8", label = T, split.by = "Condition")
DimPlot(inteData, group.by = "integrated_snn_res.1", label = T, split.by = "Condition")
```

# 4. Data exporting
```{r eval=FALSE}
saveRDS(inteData, "all_human_mouse_integration_cscChina_sample.rds")

# export the reduced results for quickly plotting
inteData@reductions$pca@assay.used <- "RNA"
inteData@reductions$umap@assay.used <- "RNA"

inteData <- DietSeurat(inteData, assays = "RNA", dimreducs = c("umap"))
saveRDS(inteData, "all_human_mouse_integration_cscChina_sample_reduced.rds")

# return to RNA assay and plotting some gene expression
DefaultAssay(inteData) <- "RNA"
inteData <- NormalizeData(inteData)
# KC related markers
FeaturePlot(inteData, features = c("KRT14", "MKI67", "KRT10", "FLG"), ncol=4, raster=T)

# Hair follicle related makers
FeaturePlot(inteData, features = c("SOX9", 'KRT75', 'KRT79', 'SCD1'), ncol=4, raster=T)

# Melanocyte and FB markers
FeaturePlot(inteData, features = c("DCT", "PDGFRA", "COL1A1", "ACTA2"), ncol=4, raster=T)

# Endothelial cell markers
FeaturePlot(inteData, features = c('PECAM1', 'CD34', "CCL21", "VWF"), ncol=4, raster=T)

# Immune cell markers
FeaturePlot(inteData, features = c("LYZ", 'CD3D', 'CD74', 'CD207'), ncol=4, raster=T)
```


############################################################
# 5. Manuscript plotting with human and mouse wound healing
```{r}
inteData <- readRDS("all_human_mouse_integration_cscChina_sample.rds")
inteData@meta.data <- droplevels(inteData@meta.data)
table(inteData$Condition, inteData$Project)
table(inteData$orig.ident, inteData$Project)
inteData$Condition <- factor(inteData$Condition, levels =
                          c("Skin","Wound1","Wound7","Wound30","YD0","YD2","YD4","YD7"))
inteData <- RunUMAP(inteData, dims = 1:30, assay = "integrated", reduction = "pca", n.neighbors = 40,
                    min.dist = 0.4, n.epochs = 400, spread = 1, learning.rate = 1)

DimPlot(inteData, group.by = "integrated_snn_res.0.8", label = T) + NoLegend()
FeaturePlot(inteData, features = c("MKI67"))
FeaturePlot(inteData, features = c("KCNK7"))

DimPlot(inteData, group.by = "orig.ident") + NoAxes()
DimPlot(inteData, group.by = "Species") + NoAxes()
DimPlot(inteData, group.by = "Project") + NoAxes()

DimPlot(inteData, group.by = "CellType", label = T, split.by = "Condition", ncol = 4, label.size = 3) + ggtitle("") + NoLegend()

DimPlot(inteData, group.by = "integrated_snn_res.0.8", label = T, split.by = "Condition", ncol = 4) + NoLegend()

FeaturePlot_scCustom(inteData, features = c("MKI67"), split.by = "Condition", num_columns =4,colors_use = c("grey90", "red"))
FeaturePlot_scCustom(inteData, features = c("FOSL1"), split.by = "Condition", num_columns =4,colors_use = c("grey90", "red"))
FeaturePlot_scCustom(inteData, features = c("S100A7"), split.by = "Condition", num_columns=4,colors_use = c("grey90", "red"))
FeaturePlot_scCustom(inteData, features = c("MMP1"), split.by = "Condition", num_columns =4,colors_use = c("grey90", "red"))
FeaturePlot_scCustom(inteData, features = c("CD68"), split.by = "Condition", num_columns =4,colors_use = c("grey90", "red"))
FeaturePlot_scCustom(inteData, features = c("EREG"), split.by = "Condition", num_columns =4,colors_use = c("grey90", "red"))

FeaturePlot_scCustom(inteData, features = c("GJB2"), split.by = "Condition", num_columns =4,colors_use = c("grey90", "red"))
FeaturePlot_scCustom(inteData, features = c("IL24"), split.by = "Condition", num_columns =4,colors_use = c("grey90", "red"))

FeaturePlot_scCustom(inteData, features = c("FOXM1"), split.by = "Condition", num_columns =4,colors_use = c("grey90", "red"))
```

```{r}
Idents(inteData) <- inteData$integrated_snn_res.1

DimPlot(inteData, group.by = "integrated_snn_res.1", label = T) + NoLegend()

DimPlot(inteData, group.by = "CellType", split.by = "Species", label = T) + NoLegend()
FeaturePlot(inteData, features = c("JCHAIN"))

Cluster_Highlight_Plot(seurat_object = inteData, cluster_name = "9", 
                       highlight_color = "forestgreen",
                       background_color = "lightgray")

DimPlot(inteData, group.by = "Species", cols = c("#d8b365", "#5ab4ac")) + ggtitle("Cross species integration") 
DimPlot(inteData, group.by = "Condition", 
        cols = c("#d89665", "#d8a965", "#d8bd65", "#d8d065",
                 "#5ab496","#63aba4", "#7dc4be", "#c9e2e0")) + ggtitle("Cross species integration") 
#pdf("Cross_species_inteUMAP_species.pdf", useDingbats = F, width = 6, height = 5)
pdf("Cross_species_inteUMAP_condition.pdf", useDingbats = F, width = 6, height = 5)
dev.off()

# re-label the cell clusters
anno.cl <- list()
anno.cl$'Bas' = c(2) 
anno.cl$'Bas_prolif' = c(19,23)
anno.cl$'Bas_mig' = c(14)
anno.cl$'Spi' = c(3,6,8)
anno.cl$'Spi_mig' = c(12) 
anno.cl$'Gra' = c(20)  
anno.cl$'HF' = c(9) 
anno.cl$'MEL' = c(25) 
anno.cl$'FB_I'  = c(0) 
anno.cl$'FB_II'  = c(13) 
anno.cl$'FB_III'  = c(15,30) 
anno.cl$'FB_prolif'  = c(22)
anno.cl$'PC_vSMC'  = c(16)
anno.cl$'LE'  = c(28) 
anno.cl$'VE'  = c(10) 
anno.cl$'Tcell'  = c(11,17)
anno.cl$'Plasma_Bcell'  = c(29)
anno.cl$'Mast_cell'  = c(21)
anno.cl$'Mono_Mac1'  = c(18)
anno.cl$'Mono_Mac2'  = c(1,24,31)
anno.cl$'Mono_Mac3'  = c(7)
anno.cl$'Neut'  = c(4)
anno.cl$'DC'  = c(5,26,27,32)

trans = rep(names(anno.cl), times = unlist(lapply(anno.cl, length)))
names(trans) = unlist(anno.cl)

inteData$newCellType = trans[as.character(inteData$integrated_snn_res.1)] %>% as.character()
inteData$newCellType = factor(inteData$newCellType, levels = names(anno.cl))
table(inteData$newCellType, inteData$Condition)


ct.cols <- c("#d94701", "#fd8d3c", "#fdbe85", #Basal clusters
             "#33A02C",  "#B2DF8A", #Spinous clusters
             "#f768a1", "#d4b9da", #Granular, Hair follicle
             "#737373", #MEL
             "#0570b0", "#3690c0", "#92c5de", "#d1edf0", #Fibroblast clusters
             "#1a9850", "#fb9a99", "#8d4720", 
             "#35978f", "#41b6c4", "#c0a700",
             "#df65b0", "#dd3497", "#6a3d9a","#80cdc1", "#b15928"
)

fact_lev <- c("Bas","Bas_prolif","Bas_mig","Spi","Spi_mig","Gra",
              "HF", "MEL",
              "FB_I","FB_II","FB_III","FB_prolif",
              "PC_vSMC","LE", "VE", 
              "Tcell","Plasma_Bcell","Mast_cell",
              "Mono_Mac1","Mono_Mac2","Mono_Mac3", "Neut", "DC")

names(ct.cols) <- fact_lev

#pdf("Cross_species_inteUMAP_celltype.pdf", useDingbats = F, width = 7, height = 6)
DimPlot(inteData, group.by = "newCellType", label = T, cols = ct.cols) +NoLegend()
#dev.off()
#pdf("Cross_species_inteUMAP_celltype_species.pdf", useDingbats = F, width = 12, height = 6)
DimPlot(inteData, group.by = "newCellType", label = T, cols = ct.cols, split.by = "Species") +NoLegend()
#dev.off()
#pdf("Cross_species_inteUMAP_celltype_species_condition.pdf", useDingbats = F, width = 18, height = 10)
DimPlot(inteData, group.by = "newCellType", label = T, cols = ct.cols, split.by = "Condition", ncol = 4) +NoLegend()
#dev.off()
```
# 5.2 Calculate the double positive cells
```{r}
genes_all <- FetchData(inteData, vars = c("Condition", "orig.ident", 
                                          "Species", "newCellType",
                                          "CD68","PTPRC","EREG","FOSL1","MKI67"), 
                       slot = "data")
data.table::fwrite(genes_all, "cross_species_expCD68_45_EREG_MKI67_FOSL1.txt", sep = "\t")

genes_human <- genes_all %>% dplyr::filter(Species == "Human")
genes_mouse <- genes_all %>% dplyr::filter(Species == "Mouse")



```


# 6. Identify and plot conserved marker genes
```{r}
inteData <- NormalizeData(inteData)
Idents(inteData) <- inteData$newCellType

# compute the conserved marker genes in human and mouse
get_conserved <- function(cluster){
  FindConservedMarkers(inteData,
                       ident.1 = cluster,
                       grouping.var = "Species",
                       only.pos = TRUE, 
                       logfc.threshold = 0.3,
                       min.diff.pct = 0.1) %>%
    rownames_to_column(var = "gene") %>% dplyr::mutate(cluster = cluster)
}

# Iterate function across desired clusters
#conserved_markers <- map_dfr(c(fact_lev), get_conserved)
#conserved_markers <- conserved_markers %>% mutate(avg_fc = (Human_avg_log2FC + Mouse_avg_log2FC) /2) %>% select(14,15,12,13,everything())
#data.table::fwrite(conserved_markers, file = "Cross_species_inteUMAP_celltype_markers.txt", sep = "\t")

conserved_markers <- data.table::fread("Cross_species_inteUMAP_celltype_markers.txt")
conserved_markers <- conserved_markers %>% group_by(cluster) %>% arrange(cluster,desc(avg_fc))
conserved_markers$cluster <- as.factor(conserved_markers$cluster)
#top200markers <- conserved_markers %>% filter(max_pval < 0.01) %>% group_by(cluster) %>% top_n(n = 200, wt = avg_fc)
#data.table::fwrite(top200markers, "Cross_species_inteUMAP_celltype_markers_top200.txt", sep = "\t")

FeaturePlot(inteData, features = c("NRG1", "IL24", "AREG", "FOSL1", "GJB2"), split.by = "Species")

# Extract top 10 markers per cluster
top10 <- conserved_markers %>% filter(max_pval < 0.01) %>% 
  group_by(cluster) %>% 
  top_n(n = 10, wt = avg_fc)

DotPlot(inteData, features = c(unique(top10$gene), "NRG1", "IL24", "FOSL1", "GJB2", "AREG"), group.by = "newCellType", cols = c("#c0a700", "#41b6c4"), 
        dot.scale = 5, col.min = 0, dot.min = 0.1, split.by = "Species"
) + theme(axis.text.x = element_text(angle = 90),
          panel.border = element_rect(colour = "black", fill=NA)) + labs(x="", y="") + coord_flip()


markergenes <- c(unique(top10$gene), "NRG1", "IL24", "FOSL1", "GJB2", "AREG")

markergenes <- c("KRT15", "MKI67", "FGFBP1", "IL24", "FOSL1",
                 "KRT1", "KRT6A", "KRT16", "LORICRIN", 
                 "SOX9", "DCT", 
                 "POSTN", "APOD", "MMP2", "PCLAF",
                 "MYH11", "ACTA2","CCL21","PECAM1",
                 "CD3D","CD3G","IGKC","IGHM","TPSB2", "GATA2",
                 "CTSS", "CD14","IL1B","CXCL3","MRC1",
                 "S100A8","G0S2","CD74","CCR7")

jjDotPlot(object = inteData,
          gene = markergenes,
          id = 'newCellType',
          rescale = T,
          rescale.min = 0,
          rescale.max = 1,
          split.by = 'Species',split.by.aesGroup = T,
          xtree = FALSE,ytree = F,
          gene.order = markergenes,
          cluster.order =  unlist(lapply(rev(levels(inteData$newCellType)), 
                                         FUN = function(x) paste0(x, " (",rev(c("Human","Mouse")), ")"))))
#pdf("Cross_species_inteUMAP_celltype_species_conservedMarkers.pdf", useDingbats = F, width = 12, height = 14)
#dev.off()
```


# 7. Pie chart of cell proportion
```{r eval=FALSE}
###############################################
#-- Cell proportion analysis across species --#
## Step 1. calculate the cell types number per sample, add the location information, 
## total numbers, as well the condition information
metadata <- inteData@meta.data %>% rownames_to_column(var = "barcode")
#data.table::fwrite(metadata, "Cross_species_inteUMAP_celltype_metadata.txt", sep = "\t")

metadata <- data.table::fread("Cross_species_inteUMAP_celltype_metadata.txt")
# only use the skin for pie chart
#metadata <- metadata %>% filter(Condition %in% c("Skin", "YD0"))
metadata <- droplevels(metadata)
cl_species <- table(metadata$Condition, metadata$Species) %>% as.data.frame() %>% dplyr::filter(Freq > 0)

sm_ct_nu <- table(metadata$Condition, metadata$newCellType) %>% as.data.frame() %>% 
  setNames(c("Sample", "CellCluster", "Num")) %>%
  left_join(., cl_species, by = c("Sample" = "Var1"))  %>% mutate(prop = Num / Freq)

## step 2. calculate the total normalized proportions of each cell type per condition
df.group <- sm_ct_nu %>% group_by(CellCluster, Var2) %>% summarise(Freq2=sum(prop)) %>% 
  ungroup() %>% group_by(CellCluster) %>% mutate(Freq_new = Freq2/sum(Freq2), lbl = scales::percent(Freq_new)) %>% ungroup()

df.group$CellCluster <- factor(df.group$CellCluster, levels = fact_lev)
ggplot(data=df.group, aes(x=" ", y=Freq_new, group=Var2, colour=Var2, fill=Var2)) +
  geom_bar(width = 1, stat = "identity") +
  scale_fill_manual(values = c("#c45027", "#00afbb")) +
  scale_color_manual(values = c("#c45027", "#00afbb")) +
  coord_polar("y", start=0) + 
  facet_grid(CellCluster ~ .) + theme_void()
```

# 8. Dynamic changes of cell proportion between human and mouse
```{r}
ct.cols <- c("#d94701", "#fd8d3c", "#fdbe85", #Basal clusters
             "#33A02C",  "#B2DF8A", #Spinous clusters
             "#f768a1", "#d4b9da", #Granular, Hair follicle
             "#737373", #MEL
             "#0570b0", "#3690c0", "#92c5de", "#d1edf0", #Fibroblast clusters
             "#1a9850", "#fb9a99", "#8d4720", 
             "#35978f", "#41b6c4", "#c0a700",
             "#df65b0", "#dd3497", "#6a3d9a","#80cdc1", "#b15928"
)

fact_lev <- c("Bas","Bas_prolif","Bas_mig","Spi","Spi_mig","Gra",
              "HF", "MEL",
              "FB_I","FB_II","FB_III","FB_prolif",
              "PC_vSMC","LE", "VE", 
              "Tcell","Plasma_Bcell","Mast_cell",
              "Mono_Mac1","Mono_Mac2","Mono_Mac3", "Neut", "DC")

names(ct.cols) <- fact_lev

metadata <- data.table::fread("Cross_species_inteUMAP_celltype_metadata.txt")
metadata$Location <- ifelse(metadata$newCellType %in% c("Bas","Bas_prolif","Bas_mig",
                                                        "Spi","Spi_mig","Gra",
                                                        "HF", "MEL"), "Epidermis", "Dermis") 
cl_species <- table(metadata$Condition, metadata$Location) %>% 
  as.data.frame() %>% setNames(c("Condition", "Location", "TotNum"))
#cl_species_hs <- cl_species %>% filter(Var1 %in% c("Skin", "Wound1", "Wound7", #"Wound30"))
#cl_species_mm1 <- cl_species %>% filter(!Var1 %in% c("Skin", "Wound1", "Wound7", #"Wound30")) %>% 
#  dplyr::filter(Var2 == "Epidermis") %>% group_by(Var1) %>% summarise(Freq=sum(Freq)) #%>% mutate(Var2 = "Epidermis") %>% ungroup() %>% select(1,3,2)
#cl_species_mm2 <- cl_species %>% filter(!Var1 %in% c("Skin", "Wound1", "Wound7", #"Wound30")) %>% 
#  group_by(Var1) %>% summarise(Freq=sum(Freq)) %>% mutate(Var2 = "Dermis") %>% ungroup() #%>% select(1,3,2)
#cl_species <- rbind(cl_species_hs, cl_species_mm1, cl_species_mm2) %>% #setNames(c("Condition", "Location", "TotNum"))

mt_df <- metadata %>% select(15,16) %>% distinct()
sm_ct_nu <- table(metadata$Condition, metadata$newCellType) %>% as.data.frame() %>% 
  setNames(c("Condition", "newCellType", "Num")) %>%
  left_join(., mt_df, by=c("newCellType")) %>% 
  left_join(., cl_species, by = c("Condition", "Location"))  %>% mutate(prop = Num / TotNum)

sm_ct_nu$Species <- ifelse(sm_ct_nu$Condition %in% c("Skin", "Wound1", "Wound7", "Wound30"), "Human", "Mouse")
sm_ct_nu$Condition <- factor(sm_ct_nu$Condition, levels = c("Skin", "Wound1", "Wound7", "Wound30",
                                                            "YD0","YD2","YD4","YD7"))
sm_ct_nu$newCellType <- factor(sm_ct_nu$newCellType, levels = fact_lev)

require(ggrepel)
ggplot(data=sm_ct_nu, aes(x=Condition, y=prop, group=newCellType, colour=newCellType)) +
  geom_line()+
  geom_point(aes(shape=newCellType)) +
  #geom_text_repel(
  #  aes(label = newCellType), data = sm_ct_nu,
  #  fontface ="plain", color = "black", size = 3
  #) +
  theme_classic() + 
  facet_wrap(vars(Species))

ggplot(data=sm_ct_nu, aes(x=Condition, y=prop, group=Species, colour=Species)) +
  geom_line()+
  geom_point(aes(shape=Species)) +
  theme_classic() + 
  facet_wrap(vars(newCellType))

# remove the Bas_mig cluster
fact_lev <- fact_lev[-3]
plot_list <- list()
for (i in seq_along(fact_lev)) {
  df_tmp <- sm_ct_nu %>% filter(newCellType == fact_lev[i])
  maxlim <- max(df_tmp$prop) + 0.001
  plot_list[[i]] <- ggplot(data=df_tmp, aes(x=Condition, y=prop, group=Species, colour=Species)) +
                           geom_line()+
                           geom_point(aes(shape=Species)) +
                           scale_y_continuous(limits = c(0, maxlim)) +
                           theme_classic() + 
                           facet_wrap(vars(newCellType))
}
wrap_plots(plot_list, ncol = 6, guides = "collect")

pdf("Cross_species_inteUMAP_celltype_species_cellproportion_changeKC.pdf", useDingbats = FALSE, width = 15, height = 10) 
dev.off()
```

# 9. Cell cycle analysis
```{r}
inteData <- NormalizeData(inteData)
inteData <- CellCycleScoring(inteData, 
                             s.features = cc.genes.updated.2019$s.genes, 
                             g2m.features = cc.genes.updated.2019$g2m.genes, set.ident = F)

p.umap.kera <- DimPlot(inteData, group.by = "Phase", label = F) +
  NoAxes() + ggtitle("")
p.umap.kera

#pdf("Cross_species_inteUMAP_cellcycle_phase.pdf", useDingbats = F, width = 8, height = 4)
DimPlot(inteData, group.by = "Phase", label = F, split.by = "Species", cols = c('grey90','#ff7f00','#33a02c'), pt.size = 0.1) +
  NoAxes() + ggtitle("")
#dev.off()

#pdf("Cross_species_inteUMAP_cellcycle_phase_condition.pdf", useDingbats = F, width = 14, height = 7)
DimPlot(inteData, group.by = "Phase", label = F, split.by = "Condition", ncol = 4, cols = c('grey90','#ff7f00','#33a02c'), pt.size = 0.1) +
  NoAxes() + ggtitle("")
#dev.off()
```

# 10. Mouse panniculus carnosus muscle and migration and proliferation scores
```{r}
# Check the marker gene expression of mouse paniculus carnosus 
FeaturePlot(inteData, features = c("MYF5","MYF6", "PAX3", "PAX7", "MYH1"), order = T, cols = c("grey90", "red"), ncol = 4)
#> Warning message:
#> The following requested variables were not found: MYF5, MYF6, PAX7, MYH1


# Conserved markers between human and mouse
conserved_markers_anno <- data.table::fread("Cross_species_inteUMAP_celltype_markers_top200.txt")
conserved_markers_anno$cluster <- as.factor(conserved_markers_anno$cluster)

# Extract top 10 markers per cluster
top10 <- conserved_markers_anno %>% 
  group_by(cluster) %>% 
  top_n(n = 10, wt = avg_fc)
basmig <- top10 %>% dplyr::filter(cluster == "Bas_mig") %>% pull(gene)
basmig <- list(basmig = basmig)
Spimig <- top10 %>% dplyr::filter(cluster == "Spi_mig") %>% pull(gene)
Spimig <- list(Spimig = Spimig)
Prolif <- top10 %>% dplyr::filter(cluster == "Bas_prolif") %>% pull(gene)
Prolif <- list(Prolif = Prolif)
# add the migration score (GO:0051549) to meta.data (KET16 AND KRT6A/B/C from literatures)
mig_gene <- list(mig_score=c("HAS2", "FGF7", "KRT16", "KRT6A", "KRT6B", "KRT6C", 
                             "MAPRE2", "IQSEC1", "FGF10", "ADAM9", "MMP9", "ARF6",
                             "IQSEC1", "MTOR", "MAP4K4", "EPB41L4B", "HBEGF"))
# combine all the migrating genes
mig_com <- list(mig_score= c(mig_gene$mig_score, basmig$basmig, Spimig$Spimig))

# add the proliferation score(Science, 374 (6574), abe6474. • DOI: 10.1126/science.abe6474)
prolif_gene <- list(prolif_score=c("ZWINT", "E2F1", "FEN1", "FOXM1", "H2AZ1", "HMGB2", 
                                   "MCM2", "MCM3", "MCM4", "MCM5", "MCM6", "MKI67", "MYBL2", 
                                   "PCNA", "PLK1", "CCND1", "AURKA", "BUB1", "TOP2A", "TYMS", 
                                   "DEK", "CCNB1", "CCNE1"))
# combine all the proliferating genes
prolif_com <- list(prolif_score= c(prolif_gene$prolif_score, Prolif$Prolif))

inteData <- AddModuleScore(inteData, features = mig_gene, name = "mig_score", assay = "RNA") 
inteData <- AddModuleScore(inteData, features = prolif_gene,name="prolif_score",assay = "RNA") 
inteData <- AddModuleScore(inteData, features = basmig,name="BasMig_score", assay = "RNA") 
inteData <- AddModuleScore(inteData, features = Spimig,name= "Spimig_score",assay = "RNA") 
inteData <- AddModuleScore(inteData, features = Prolif,name="newProlif_score",assay = "RNA") 
inteData <- AddModuleScore(inteData, features = mig_com,name= "mig_com_score",assay = "RNA") 
inteData <- AddModuleScore(inteData, features = prolif_com,name="prolif_com_score",assay = "RNA") 


VlnPlot(inteData, features = "mig_score1", pt.size = 0, group.by = 'newCellType')
VlnPlot(inteData, features = "prolif_score1", pt.size = 0, group.by = 'newCellType')
VlnPlot(inteData, features = "BasMig_score1", pt.size = 0.0001, group.by = 'newCellType')
VlnPlot(inteData, features = "Spimig_score1", pt.size = 0.0001, group.by = 'newCellType')
VlnPlot(inteData, features = "newProlif_score1", pt.size = 0.0001, group.by = 'newCellType')
VlnPlot(inteData, features = "mig_com_score1", pt.size = 0.0001, group.by = 'newCellType')
VlnPlot(inteData, features = "prolif_com_score1", pt.size = 0.0001, group.by = 'newCellType')

mt_scores <- inteData@meta.data %>% dplyr::filter(newCellType == "Bas_mig" | newCellType == "Spi_mig" | newCellType == "Bas_prolif")
mt_scores$Condition <- factor(mt_scores$Condition, levels = c("Skin", "Wound1", "Wound7", "Wound30",
                                                              "YD0","YD2","YD4","YD7"))

p1 <- ggplot(mt_scores, aes(x=BasMig_score1, y=newProlif_score1,
                            color=newCellType, fill=newCellType)) +
  geom_point(size=0.8) +
  #scale_color_manual(values=nodecolor)+ 
  scale_x_continuous(limits = c(-0.2,2.2),breaks = c(0, 0.5,1,1.5,2)) +
  scale_y_continuous(limits = c(-0.12, 0.8),breaks = c(0, 0.2, 0.4,0.6,0.8)) +
  xlab("Basal migration score") + ylab("Proliferation score") + 
  theme_classic() +
  theme(
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank(),
    panel.border = element_rect(colour = "black",fill = NA,size = 1),
    axis.text.x = element_text(size = 12), 
    axis.text.y = element_text(size = 12),
    axis.title.x = element_text(size = 14),
    axis.title.y = element_text(size = 14),
    legend.position = "right",#top,bottom,left,right
    legend.spacing = unit(0,"cm"),
    aspect.ratio = 1) +
  facet_wrap(vars(Condition), ncol = 4)
p1

p2 <- ggplot(mt_scores, aes(x=Spimig_score1, y=newProlif_score1,
                            color=newCellType, fill=newCellType)) +
  geom_point(size=0.8) +
  #scale_color_manual(values=nodecolor)+ 
  #scale_x_continuous(limits = c(-0.2,2.2),breaks = c(0, 0.5,1,2)) +
  scale_y_continuous(limits = c(-0.12, 0.8),breaks = c(0, 0.2, 0.4,0.6,0.8)) +
  xlab("Spinous migration score") + ylab("Proliferation score") + 
  theme_classic() +
  theme(
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank(),
    panel.border = element_rect(colour = "black",fill = NA,size = 1),
    axis.text.x = element_text(size = 12), 
    axis.text.y = element_text(size = 12),
    axis.title.x = element_text(size = 14),
    axis.title.y = element_text(size = 14),
    legend.position = "right",#top,bottom,left,right
    legend.spacing = unit(0,"cm"),
    aspect.ratio = 1) +
  facet_wrap(vars(Condition), ncol = 4)
p2

p1/p2 + plot_layout(guides = "collect")


p3 <- ggplot(mt_scores, aes(x=mig_score1, y=prolif_score1,
                            color=newCellType, fill=newCellType)) +
  geom_point(size=0.8) +
  #scale_color_manual(values=nodecolor)+ 
  #scale_x_continuous(limits = c(-0.2,2.2),breaks = c(0, 0.5,1,2)) +
  #scale_y_continuous(limits = c(-0.12, 0.8),breaks = c(0, 0.2, 0.4,0.6,0.8)) +
  xlab("GO migration score") + ylab("GO proliferation score") + 
  theme_classic() +
  theme(
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank(),
    panel.border = element_rect(colour = "black",fill = NA,size = 1),
    axis.text.x = element_text(size = 12), 
    axis.text.y = element_text(size = 12),
    axis.title.x = element_text(size = 14),
    axis.title.y = element_text(size = 14),
    legend.position = "right",#top,bottom,left,right
    legend.spacing = unit(0,"cm"),
    aspect.ratio = 1) +
  facet_wrap(vars(Condition), ncol = 4)
p3

p4 <- ggplot(mt_scores, aes(x=mig_com_score1, y=prolif_com_score1,
                            color=newCellType, fill=newCellType)) +
  geom_point(size=0.8) +
  #scale_color_manual(values=nodecolor)+ 
  #scale_x_continuous(limits = c(-0.2,2.2),breaks = c(0, 0.5,1,2)) +
  #scale_y_continuous(limits = c(-0.12, 0.8),breaks = c(0, 0.2, 0.4,0.6,0.8)) +
  xlab("Combined migration score") + ylab("Combined proliferation score") + 
  theme_classic() +
  theme(
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank(),
    panel.border = element_rect(colour = "black",fill = NA,size = 1),
    axis.text.x = element_text(size = 12), 
    axis.text.y = element_text(size = 12),
    axis.title.x = element_text(size = 14),
    axis.title.y = element_text(size = 14),
    legend.position = "right",#top,bottom,left,right
    legend.spacing = unit(0,"cm"),
    aspect.ratio = 1) +
  facet_wrap(vars(Condition), ncol = 4)
p4

p3/p4 + plot_layout(guides = "collect")
```


# 11. DE analysis between the marker 
```{r}
# DE analysis between human and mouse data
Idents(inteData) <- inteData$newCellType
# load the gene annotation file
anno_human <- readRDS("/Users/zhuliu/Desktop/sc st/Versions_SC_ST_Project/V2_Figures_231017/01-Scripts/custom functions/humanAnnotation.rds")

####----DE functions----####
DE.genes <- function(group1 = NULL, group2 = NULL, groupby = NULL, subsetIdent = NULL, minPct = 0.25, fc = 0){
  tmp <- FindMarkers(inteData, 
                     ident.1 = group1, 
                     ident.2 = group2, 
                     group.by = groupby, 
                     subset.ident = subsetIdent,
                     assay = "RNA",
                     test.use = "MAST",
                     min.pct = minPct, 
                     logfc.threshold = fc,
                     min.diff.pct = 0.3)
  tmp <- tmp %>% tibble::rownames_to_column(var = "gene") %>% 
    arrange(desc(avg_log2FC)) %>% 
    left_join(., anno_human, by=c("gene" = "external_gene_name")) %>% 
    dplyr::select(1, 3, 2, 6, everything()) %>% 
    #filter(p_val_adj < 0.05) %>% #or filter out the genes with p-value < 0.05
    mutate(Type = factor(ifelse(p_val_adj < 0.05 & abs(avg_log2FC) >= 0.3, ifelse(avg_log2FC >= 0.3 ,'Up','Down'),'Not'))) #add a column with regulation direction
  return(tmp)
}

##-- draw the volcano plot --##
DE.volcano <- function(data = NULL, avgfc = 1, padj = 0.05){
  require(ggrepel)
  data.fil <- data %>% filter(abs(avg_log2FC) >= avgfc & p_val_adj < padj)
  ggplot(data=data, aes(x=avg_log2FC, y=-log10(p_val_adj), color=Type, fill=Type)) +
    geom_point(size=1.5)+
    scale_color_manual(values=c("blue","#B3B3B3", "red"))+
    coord_fixed()+
    theme_classic() +
    theme(
      legend.position = "right",#top,bottom,left,right
      legend.spacing = unit(0,"cm"),
      aspect.ratio = 1) +
    geom_text_repel(data = data.fil, mapping = aes(x=avg_log2FC, y=-log10(p_val_adj), label = gene), show.legend = FALSE, max.overlaps = 50)
}

celltypename = "Bas_mig"
DE.genes.sorted <- DE.genes(group1 = "Human", group2 = "Mouse", groupby = "Species", subsetIdent = celltypename)
DE.genes.sorted.fil <- DE.genes.sorted %>% filter(abs(avg_log2FC) > 1 & p_val_adj < 0.05)
DE.volcano(data = DE.genes.sorted, avgfc = 2)

celltypename = "Spi_mig"
DE.genes.sorted2 <- DE.genes(group1 = "Human", group2 = "Mouse", groupby = "Species", subsetIdent = celltypename)
DE.genes.sorted2.fil <- DE.genes.sorted2 %>% filter(abs(avg_log2FC) > 1 & p_val_adj < 0.05)
DE.volcano(data = DE.genes.sorted2, avgfc = 2)

# Combined cluster Bas_mig and Spi_mig
celltypename = c("Bas_mig", "Spi_mig")
DE.genes.sorted3 <- DE.genes(group1 = "Human", group2 = "Mouse", groupby = "Species", subsetIdent = celltypename)
DE.genes.sorted3.fil <- DE.genes.sorted3 %>% filter(abs(avg_log2FC) > 1 & p_val_adj < 0.05)
DE.volcano(data = DE.genes.sorted3, avgfc = 2)

# DEGs venn Diagram
library(ggVennDiagram)
library(ggplot2)
gene_list <- list(Bas_up = DE.genes.sorted.fil %>% 
                    dplyr::filter(Type == "Up") %>% pull(gene),
                  Spi_up = DE.genes.sorted2.fil %>% 
                    dplyr::filter(Type == "Up") %>% pull(gene))
p1 <- ggVennDiagram(gene_list)
p1
gene_list <- list(Bas_down = DE.genes.sorted.fil %>% 
                    dplyr::filter(Type == "Down") %>% pull(gene),
                  Spi_down = DE.genes.sorted2.fil %>% 
                    dplyr::filter(Type == "Down") %>% pull(gene))
p2 <- ggVennDiagram(gene_list)
p2
p1 + p2
#pdf("DEGs_overlapNumber.pdf", useDingbats = F, width = 8, height = 6)
#dev.off()

#################################
#- Extract the gene expression -#
#genes_exp <- AggregateExpression(inteData,assays = "RNA", features = DE.genes.sorted.fil$gene, slot = "data")$RNA
genes_exp1 <- FetchData(inteData, vars = DE.genes.sorted2.fil$gene, slot = "data", cells = WhichCells(inteData, ident = "Bas_mig"))
genes_exp2 <- FetchData(inteData, vars = DE.genes.sorted2.fil$gene, slot = "data", cells = WhichCells(inteData, ident = "Spi_mig"))
# maximun expression values
max(genes_exp1);max(genes_exp2)
min(genes_exp1);min(genes_exp2)
gene1 <- apply(genes_exp1, 2, max) %>% as.data.frame() %>% rownames_to_column(var = "genes") %>% setNames(c("gene", "exp"))
gene2 <- apply(genes_exp2, 2, max) %>% as.data.frame() %>% rownames_to_column(var = "genes") %>% setNames(c("gene", "exp"))
# only keep the maximum expression
gene_max <- gene1 %>% left_join(., gene2, by="gene")
gene_max$expmax <- apply(gene_max[, 2:3], 1, max)

# make the new data for volcano plot
DE.genes.sorted.fil.new <- DE.genes.sorted3.fil %>% left_join(., gene_max[, c(1,4)], by= c("gene"))

# highlight the interested genes
## according to fold change 
hig_fc <- c("S100A2", "S100A7", "S100A8", "S100A9", "F12", "TNFSF10", "IFI16", "MMP1", "SERPINB4", "SERPINB3")
FeaturePlot(inteData, features =hig_fc[1:5], split.by = "Condition", cols = c("grey90", "red"), order = F)
FeaturePlot(inteData, features =hig_fc[6:10], split.by = "Condition", cols = c("grey90", "red"), order = F)
## according to expression
hig_mouse <- c("PABPC3", "LGALS7", "TMSB4Y", "MT1A", "TPM2", "PGAM4", "RBP2", "PI15", "TRBC2", "SPRR1A")
FeaturePlot(inteData, features =hig_mouse[1:5], split.by = "Condition", cols = c("grey90", "red"), order = F)
FeaturePlot(inteData, features =hig_mouse[6:10], split.by = "Condition", cols = c("grey90", "red"), order = F)

DE.genes.sorted.fil.hig1 <- DE.genes.sorted.fil.new %>% 
  arrange(desc(avg_log2FC)) %>% dplyr::slice(1:30)
DE.genes.sorted.fil.hig2 <- DE.genes.sorted.fil.new %>% 
  arrange(avg_log2FC, desc(expmax)) %>% dplyr::slice(1:30)
FeaturePlot(inteData, features =DE.genes.sorted.fil.hig2$gene[1:5], split.by = "Condition", cols = c("grey90", "red"), order = F)

#DE.genes.sorted.fil.hig3 <- DE.genes.sorted.fil.new %>% filter(gene %in% c("KRT16", "KRT17", "KRT6A", "GJB2", "IL18", "IFI16"))
DE.genes.sorted.fil.hig <- DE.genes.sorted.fil.new %>% filter(gene %in% c(hig_fc, hig_mouse))

FeaturePlot(inteData, features =c("MKI67"), split.by = "Condition", cols = c("grey90", "red"), order = F)

#pdf("Cross_species_inteUMAP_cellcycle_phase_MKI67.pdf", useDingbats = F, width = 14, height = 7)
FeaturePlot_scCustom(seurat_object = inteData, features = "MKI67", order = T, split.by = "Condition", num_columns = 4) & NoAxes()
#dev.off()

pdf("Cross_species_inteUMAP_DEGs_S100family.pdf", useDingbats = F, width = 8, height = 21)
FeaturePlot(inteData, 
            features =c("S100A2", "S100A7", "S100A8", "S100A9","MMP1", "SERPINB3","SERPINB4"), 
            split.by = "Condition", 
            cols = c("grey90", "red"), order = F) 
FeaturePlot_scCustom(seurat_object = inteData, features = c("S100A2", "S100A7", "S100A8", "S100A9","MMP1", "SERPINB3","SERPINB4"), order = F, split.by = "Condition", num_columns = 8) & NoAxes()
FeaturePlot_scCustom(seurat_object = inteData, features = c("S100A2", "S100A7", "S100A8", "S100A9","MMP1", "SERPINB3","SERPINB4"), order = F, split.by = "Species", num_columns = 2) & NoAxes()
dev.off()

#pdf("DEGs_migrationGenes_noOrder.pdf", useDingbats = F, width = 8, height = 15)
FeaturePlot_scCustom(seurat_object = inteData, features = c("NRG1", "IL24", "AREG", "FOSL1","GJB2"), order = F, split.by = "Condition", num_columns = 8) & NoAxes()
FeaturePlot_scCustom(seurat_object = inteData, features = c("NRG1", "IL24", "AREG", "FOSL1","GJB2"), order = F, split.by = "Species") & NoAxes()
#dev.off()

ggplot(data=DE.genes.sorted.fil.new, aes(x=avg_log2FC, y=expmax, color=Type, fill=Type)) +
    geom_point(size=1.5)+
    scale_color_manual(values=c("blue", "red"))+
    coord_fixed() +
    labs(x = "log2(fold change)", y="Migrating cluster3&4 maximum expression") +
    geom_vline(xintercept=c(-1,1), lty=4, col="grey", lwd=0.6)+ 
    #geom_hline(yintercept = -log10(0.05), lty=4, col="grey", lwd=0.6)+
    #scale_x_continuous(limits = c(-5,5),breaks = c(-4,-3,-2,-1,0,1,2,3,4)) +
    scale_y_continuous(limits = c(0,8),breaks = c(0,2,4,6,8,10)) +
    theme_classic() +
    theme(
      panel.grid.major = element_line(colour = "grey90",size = 0.5,linetype = 2),
      panel.grid.minor = element_line(colour = "grey90",size = 0.5,linetype = 2),
      axis.text.x = element_text(size = 12), 
      axis.text.y = element_text(size = 12),
      axis.title.x = element_text(size = 14),
      axis.title.y = element_text(size = 14),
      legend.position = "right",#top,bottom,left,right
      legend.spacing = unit(0,"cm"),
      aspect.ratio = 1) +
    geom_text_repel(data = DE.genes.sorted.fil.hig, mapping = aes(x=avg_log2FC, y=expmax, label = gene), show.legend = FALSE, max.overlaps = 50,min.segment.length = 0.01)

#data.table::fwrite(DE.genes.sorted.fil.new, file = "DEGs_migratingKC_cluster3.txt", sep = "\t")
#data.table::fwrite(DE.genes.sorted.fil.new, file = "DEGs_migratingKC_cluster4.txt", sep = "\t")
#data.table::fwrite(DE.genes.sorted.fil.new, file = "DEGs_migratingKC_cluster3and4.txt", sep = "\t")
#pdf("DEGs_migratingKC_volcano_cluster3.pdf", useDingbats = F, width = 6, height = 6)
#pdf("DEGs_migratingKC_volcano_cluster4.pdf", useDingbats = F, width = 6, height = 6)
#pdf("DEGs_migratingKC_volcano_cluster3and4.pdf", useDingbats = F, width = 6, height = 6)
#dev.off()
```


# 12. GO analysis of DEGs
```{r}
datatmp <- DE.genes.sorted.fil.new %>% filter(Type == "Down")
require(clusterProfiler) #loading the package if it is not loaded
require(org.Hs.eg.db)
Enriched_gene_up <- as.character(unique(datatmp$entrezgene_id))
ego_BP_up <- enrichGO(gene = Enriched_gene_up,
                      OrgDb         = org.Hs.eg.db,
                      ont           = "BP",
                      keyType = "ENTREZID",
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff  = 0.1,
                      minGSSize = 10,
                      maxGSSize = 500,
                      readable = T)

go_df <- as.data.frame(ego_BP_up)
#data.table::fwrite(go_df, file = "DEGs_migratingKC_cluster3and4_Up_humanGO.txt", sep = "\t")
#data.table::fwrite(go_df, file = "DEGs_migratingKC_cluster3and4_Down_mouseGO.txt", sep = "\t")

dotplot(ego_BP_up, showCategory=20, label_format = 100) +
  theme(
    panel.grid.minor = element_blank(),
    panel.border = element_rect(size = 0.5, color = "black"),
    axis.text.x = element_text(size = 14, color = "black"),
    axis.text.y = element_text(size = 14, color = "black"))

## remove redundent GO terms
ego2 <- simplify(ego_BP_up)
cnetplot(ego2, foldChange=Enriched_gene_up)
cnetplot(ego2, foldChange=Enriched_gene_up, circular = TRUE, colorEdge = TRUE)

#pdf("DEGs_migratingKC_cluster4_Up_humanGO.pdf", useDingbats = F, width = 8, height = 6)
#pdf("DEGs_migratingKC_cluster4_Down_mouseGO.pdf", useDingbats = F, width = 8, height = 6)
#dev.off()


go1 <- readxl::read_xlsx("/Users/zhuliu/Desktop/sc st/Versions_SC_ST_Project/V3_Figures_240226/Table S3 DEGs_GO_cross species.xlsx", sheet = 1) %>% dplyr::slice(1:6) %>% mutate(species = "Human")
go2 <- readxl::read_xlsx("/Users/zhuliu/Desktop/sc st/Versions_SC_ST_Project/V3_Figures_240226/Table S3 DEGs_GO_cross species.xlsx", sheet = 2) %>% dplyr::slice(1:6) %>% mutate(species = "Mouse")

go_plot <- rbind(go1, go2)

go_plot$Description <- factor(go_plot$Description, levels = rev(go_plot$Description))
ggplot(data = go_plot, aes(x=Description, y = -log2(qvalue), 
                           color = species)) + 
  scale_color_manual(values = c("#c45027", "#00afbb")) +
  geom_point(stat='identity', fill="black", size=6) +
  geom_segment(aes(y=0, x=Description, yend=-log2(qvalue), xend=Description)) +
  coord_flip() +
  #scale_y_continuous(limits = c(0,30),breaks = c(0,10,20,30)) +
  theme_bw() + 
  ylab("") + 
  xlab("") +
  theme(axis.text.x = element_text(size = 12, angle = 0, hjust = 1))
#pdf("DEGs_migratingKC_GO_BPs_cluster3and4_NEWavalue.pdf", useDingbats = F, width = 6, height = 4)
#dev.off()
```


# SessionInfo
```{r}
sessionInfo()
```

