---
title: "Nichenet ligand-receptor analysis"
author: "Zhuang Liu"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true #add a table of contents (TOC)
    toc_depth: 3 #headers' level in the toc
    toc_float: true #toc visible even when scrolled
    theme: lumen
    highlight: tango
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align='center', fig.width=12, fig.height = 8)
```

```{r warning=FALSE, message=FALSE}
# add the local lib path
.libPaths(c("/proj/snic2021-23-156/proj_10XpostAnalysis/multinichenet/library" , .libPaths()))
library(tidyverse)
library(SingleCellExperiment)
library(dplyr)
library(ggplot2)
library(multinichenetr)
```

####################################################
# Step 0. Prepare NicheNet analysis
####################################################
## 1. Load the required Ligand-target model
```{r}
organism = "human"
if(organism == "human"){
  lr_network = readRDS("new2022_db/lr_network_human_21122021.rds")
  lr_network = lr_network %>% dplyr::rename(ligand = from, receptor = to) %>% distinct(ligand, receptor) %>% mutate(ligand = make.names(ligand), receptor = make.names(receptor))
  ligand_target_matrix = readRDS("new2022_db/ligand_target_matrix_nsga2r_final.rds")
  colnames(ligand_target_matrix) = colnames(ligand_target_matrix) %>% make.names()
  rownames(ligand_target_matrix) = rownames(ligand_target_matrix) %>% make.names()
} else if(organism == "mouse"){
  lr_network = readRDS("new2022_db/lr_network_mouse_21122021.rds")
  lr_network = lr_network %>% dplyr::rename(ligand = from, receptor = to) %>% distinct(ligand, receptor) %>% mutate(ligand = make.names(ligand), receptor = make.names(receptor))
  ligand_target_matrix = readRDS("new2022_db/ligand_target_matrix_nsga2r_final_mouse.rds")
  colnames(ligand_target_matrix) = colnames(ligand_target_matrix) %>% make.names()
  rownames(ligand_target_matrix) = rownames(ligand_target_matrix) %>% make.names()
}
```

## 2. read in SingleCellExperiment Objects
```{r}
seurat_obj <- readRDS(file = "./allcombined_wounds_newAnnotation.rds")
table(seurat_obj$newCellTypes, seurat_obj$Condition)
table(seurat_obj$newCellTypes, seurat_obj$orig.ident)

# load the clean metadata
mt <- data.table::fread("./Subclusters.txt")
mt$CellType <- gsub("\\/|-", "_", mt$CellType)

seurat_obj <- seurat_obj[, mt$barcode]

identical(colnames(seurat_obj), mt$barcode)
seurat_obj$CellType <- mt$CellType
seurat_obj <- Seurat::NormalizeData(seurat_obj)
seurat_obj <- Seurat::DietSeurat(seurat_obj, assays = "RNA")

sce = Seurat::as.SingleCellExperiment(seurat_obj, assay = "RNA")
rm(seurat_obj);invisible(gc())

# Check the gene symbol names
sce = alias_to_symbol_SCE(sce, "human") %>% makenames_SCE()

table(SummarizedExperiment::colData(sce)$CellType, SummarizedExperiment::colData(sce)$orig.ident) # cell types vs samples

table(SummarizedExperiment::colData(sce)$CellType, SummarizedExperiment::colData(sce)$Condition) # cell types vs conditions
```

## 3. Prepare settings of the MultiNicheNet cell-cell communication analysis
```{r}
sample_id = "orig.ident"
group_id = "Condition"
celltype_id = "CellType"
covariates = NA
batches = NA

senders_oi = SummarizedExperiment::colData(sce)[,celltype_id] %>% unique()
receivers_oi = SummarizedExperiment::colData(sce)[,celltype_id] %>% unique()

sce = sce[, SummarizedExperiment::colData(sce)[,celltype_id] %in% c(senders_oi, receivers_oi)]
```


###########################################
# Step 1. Abundance and expression analysis
###########################################

Extract cell type abundance and expression information from receiver and sender cell types, and link this expression information for ligands of the sender cell types to the corresponding receptors of the receiver cell types

```{r}
min_cells = 10
```


#############################################
# Step 2. DE analysis between the conditions
#############################################

Perform genome-wide differential expression analysis of receiver and sender cell types to define DE genes between the conditions of interest. Based on this analysis, we can define the logFC/p-value of ligands in senders and receptors in receivers, and define the set of affected target genes in the receiver.

```{r}
# Define the contrasts and covariates of interest for the DE analysis.
contrasts_oi = c("'Skin-(Wound1+Wound7+Wound30)/3','Wound1-(Skin+Wound7+Wound30)/3','Wound7-(Skin+Wound1+Wound30)/3','Wound30-(Skin+Wound1+Wound7)/3'")

contrast_tbl = tibble(contrast = c("Skin-(Wound1+Wound7+Wound30)/3","Wound1-(Skin+Wound7+Wound30)/3","Wound7-(Skin+Wound1+Wound30)/3","Wound30-(Skin+Wound1+Wound7)/3"), 
                      group = c("Skin","Wound1","Wound7","Wound30"))
```


###########################################################
# Step 3. Predict ligand activities and ligand-target links
###########################################################
```{r}
# Define the parameters for the NicheNet ligand activity analysis
logFC_threshold = 0.50
p_val_threshold = 0.05
fraction_cutoff = 0.05

# p_val_adj = TRUE 
p_val_adj = FALSE 
empirical_pval = FALSE

top_n_target = 250

cores_system = 15
n.cores = min(cores_system, union(senders_oi, receivers_oi) %>% length()) # use one core per receiver cell type
```


################################################################
# Step 4. Prioritize all sender-ligand - receiver-receptor pairs
################################################################

```{r}
# Define the prioritization weights, and prepare grouping objects
prioritizing_weights_DE = c("de_ligand" = 1, "de_receptor" = 1)
prioritizing_weights_activity = c("activity_scaled" = 2)

prioritizing_weights_expression_specificity = c("exprs_ligand" = 2, "exprs_receptor" = 2)

prioritizing_weights_expression_sufficiency = c("frac_exprs_ligand_receptor" = 1)

prioritizing_weights_relative_abundance = c( "abund_sender" = 0, "abund_receiver" = 0)

prioritizing_weights = c(prioritizing_weights_DE, 
                         prioritizing_weights_activity, 
                         prioritizing_weights_expression_specificity,
                         prioritizing_weights_expression_sufficiency, 
                         prioritizing_weights_relative_abundance)

multinichenet_output = multi_nichenet_analysis(sce = sce, celltype_id = celltype_id, sample_id = sample_id, group_id = group_id, 
                                lr_network = lr_network, ligand_target_matrix = ligand_target_matrix, contrasts_oi = contrasts_oi, contrast_tbl = contrast_tbl, batches = batches, covariates = covariates,
                                prioritizing_weights = prioritizing_weights, min_cells = min_cells, logFC_threshold = logFC_threshold, p_val_threshold = p_val_threshold,  
                                fraction_cutoff = fraction_cutoff, p_val_adj = p_val_adj, empirical_pval = empirical_pval, top_n_target = top_n_target, n.cores = n.cores, sender_receiver_separate = FALSE, verbose = TRUE)

saveRDS(multinichenet_output, file = "multinichenet_output.rds")
```
